{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running Tximport and DESeq2 to calculate differentially expressed genes between HER2+ and TNBC BC samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Review our pipeline\n",
    "In the pipeline image, the red text indicates the names of the R packages we will use to perform the analysis, the yellow boxes indicate the data (raw or processed), and the grey indicates what we will do in each step. Note that we will use [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) to identify differentially expressed genes (DEGs). This is a highly cited package that uses a negative binomial distribution to calculate DEGs. However, there are multiple pipelines and programs to do this. [Knoweng](http://education.knoweng.org/sequenceng/) has a nice interface to explore other options, including the strengths and weaknesses of each.\n",
    "![](https://github.com/davidnboone/test-image/blob/master/overview-image.jpg?raw=true \"pipeline overview\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To this point we have:\n",
    "1. Extracted and calculated HER2 and TNBC status of each patient.\n",
    "2. Extracted count and abundance data for each transcript from those samples.\n",
    "3. Prepped the transcipt count and abundance data for Tximport.\n",
    "\n",
    "## In this notebook we will:\n",
    "1. Run Tximport, which will collapse count and abundance data to the gene level.\n",
    "2. Calculate differentially expressed genes using DESeq2.\n",
    "\n",
    "___\n",
    "___\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We still need the proper R packages loaded. We will need to start by reloading them with the same code as in the previous Notebook.\n",
    "Of course you can use the typical library() command to load each as well.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: pacman\n"
     ]
    }
   ],
   "source": [
    "#you can very easily use \"install.packages\" or \"biocLite\" to install the packages and \"library\" to load them\n",
    "#however, instead I am installing a package called pacman that will determine if a package is already installed\n",
    "#if it is not it will install it and after will load it\n",
    "#this method is convenient when sharing code that requires the use of others packages\n",
    "\n",
    "if (!require(\"pacman\")) install.packages(\"pacman\")\n",
    "pacman::p_load(R.utils, data.table, tximport, DESeq2, biomaRt, jsonlite, BiocParallel, ggplot2, gplots, RColorBrewer, devtools, pheatmap)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set the working directory to the workspace set up in the previous notebook.\n",
    " You can check to deterimine if you are in the correct directory by examining your working directory. If it is not the proper workspace then change to proper directory.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'/Users/dnb14/Documents/jupyter/tcga_brca_rnaseq'"
      ],
      "text/latex": [
       "'/Users/dnb14/Documents/jupyter/tcga\\_brca\\_rnaseq'"
      ],
      "text/markdown": [
       "'/Users/dnb14/Documents/jupyter/tcga_brca_rnaseq'"
      ],
      "text/plain": [
       "[1] \"/Users/dnb14/Documents/jupyter/tcga_brca_rnaseq\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#only need to run if you are not in the proper working directory.\n",
    "base_dir <- \"/Volumes/Oesterreich\"\n",
    "setwd(base_dir)\n",
    "\n",
    "project_name <- \"TCGA_Her2_TNBC_DEGs\"\n",
    "date <- \"2019_02_06\" #use the same date as notebook 1 and 2\n",
    "setwd(project_name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "___\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Collapse transcript-level data to gene level using [Tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html).\n",
    "If you remember from Notebook 2, DESeq2 calculates differential *gene* expression not *transcript* expression. There are packages that will identify differentially expressed exons or transcripts, however these are still largely unreliable because many isoforms share the same exons, making precise mapping difficult. Accordinly, we need to collapse the transcript counts to gene counts. Tximport is a package that can do this for us. I highly recommend reading the [manual](https://bioconductor.org/packages/release/bioc/html/tximport.html) before starting. \n",
    "\n",
    "Tximport requires:\n",
    "1. counts or tpm data (or output from pseudoaligners Salmon or Kallisto)\n",
    "2. knowledge file that maps each transcript to gene. \n",
    "\n",
    "In notebook 2, we wrangled the count and tpm data into the proper format and created the knowledge file. Now we will run Tximport.\n",
    "\n",
    "___\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1  Read in all of the appropriate  datafiles generated in notebook 1 and 2. \n",
    "If you compiled all notebooks into one, and you have the following in your global envirenment, 'brca.clin.selected', 'brca.untran.counts.df', 'brca.untran.tpm.df', 'date', and 'tx2gene81' you can skip the next block of code and go to 1.2 Run Tximport.\n",
    "\n",
    "As explained above For tximport we need \n",
    "1. the unlogged transcript counts data\n",
    "2. the unlogged transcript tpm data\n",
    "3. tx2gene81 \n",
    "\n",
    "After tximport, we will need the clinical data for DESeq2\n",
    "4. brca.clin.selected\n",
    "\n",
    "You can read them in using the code below. **Be sure the commands point to the appropriate directories that contain the data**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in the unlogged transcript count data. Because it was saved as an Rda we can use the \n",
    "#load() function and it will retain the appropriate variable name\n",
    "load(file = file.path(date, \"df.unlogcounts_her2_tnbc.Rda\"))\n",
    "\n",
    "#read in the unlogged transcript tpm data.\n",
    "load(file = file.path(date, \"df.unlogtpm_her2_tnbc.Rda\"))\n",
    "\n",
    "#read in the tx2gene file\n",
    "load(file = file.path(date, \"tx2gene81.Rda\"))\n",
    "\n",
    "#read in the clinical data\n",
    "brca.clin.selected <- read.table(file = file.path(date,\"brca.clin.data.her2.tnbc.txt\"), header = TRUE, stringsAsFactors = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Run Tximport\n",
    "The resulting file of tximport will provide the **gene** count data. Again, please refer to the Tximport manual to see all the ways to run it. We are providing count and tpm data instead of the output data from a series of data from pseudoaligners, so this is slightly different than if you created your own counts files from fastq files using Salmon or Kallisto. Please, note that in this method, we create a list named txi. This list contains all the abundance data (tpm data that we provide), counts data (again from the counts dataframe generated in Notebook 2), and the length of each transcript. We only need to provide the length data if we are converting counts to tpm or vice versa, which we are not doing. Hence, we will just give dummy values of the same length. We also tell tximport that we are not calculating counts from abundance data for each value. To summarize to gene level, we then provide that list (txi) and the knowledge file tx2gene (generated in Notebook 2)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we run tximport, let's confirm that the transcript names in the counts dataframe and tx2gene match.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>transcript</th><th scope=col>gene</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENST00000456328</td><td>ENSG00000223972</td></tr>\n",
       "\t<tr><td>ENST00000450305</td><td>ENSG00000223972</td></tr>\n",
       "\t<tr><td>ENST00000488147</td><td>ENSG00000227232</td></tr>\n",
       "\t<tr><td>ENST00000619216</td><td>ENSG00000278267</td></tr>\n",
       "\t<tr><td>ENST00000473358</td><td>ENSG00000243485</td></tr>\n",
       "\t<tr><td>ENST00000469289</td><td>ENSG00000243485</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " transcript & gene\\\\\n",
       "\\hline\n",
       "\t ENST00000456328 & ENSG00000223972\\\\\n",
       "\t ENST00000450305 & ENSG00000223972\\\\\n",
       "\t ENST00000488147 & ENSG00000227232\\\\\n",
       "\t ENST00000619216 & ENSG00000278267\\\\\n",
       "\t ENST00000473358 & ENSG00000243485\\\\\n",
       "\t ENST00000469289 & ENSG00000243485\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "transcript | gene | \n",
       "|---|---|---|---|---|---|\n",
       "| ENST00000456328 | ENSG00000223972 | \n",
       "| ENST00000450305 | ENSG00000223972 | \n",
       "| ENST00000488147 | ENSG00000227232 | \n",
       "| ENST00000619216 | ENSG00000278267 | \n",
       "| ENST00000473358 | ENSG00000243485 | \n",
       "| ENST00000469289 | ENSG00000243485 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  transcript      gene           \n",
       "1 ENST00000456328 ENSG00000223972\n",
       "2 ENST00000450305 ENSG00000223972\n",
       "3 ENST00000488147 ENSG00000227232\n",
       "4 ENST00000619216 ENSG00000278267\n",
       "5 ENST00000473358 ENSG00000243485\n",
       "6 ENST00000469289 ENSG00000243485"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENST00000548312.5</th><td>10.01273 </td><td> 8.991491</td><td> 0.000000</td><td> 8.237484</td><td> 5.519675</td></tr>\n",
       "\t<tr><th scope=row>ENST00000527779.1</th><td>30.87603 </td><td>18.463510</td><td>44.311332</td><td> 0.000000</td><td>37.473682</td></tr>\n",
       "\t<tr><th scope=row>ENST00000454820.5</th><td> 0.00000 </td><td> 0.000000</td><td> 0.000000</td><td> 0.000000</td><td> 0.000000</td></tr>\n",
       "\t<tr><th scope=row>ENST00000535093.1</th><td> 0.00000 </td><td> 0.000000</td><td>15.978063</td><td> 0.000000</td><td> 0.000000</td></tr>\n",
       "\t<tr><th scope=row>ENST00000346219.7</th><td> 0.00000 </td><td> 0.000000</td><td> 7.487855</td><td>10.401091</td><td> 8.667213</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01\\\\\n",
       "\\hline\n",
       "\tENST00000548312.5 & 10.01273  &  8.991491 &  0.000000 &  8.237484 &  5.519675\\\\\n",
       "\tENST00000527779.1 & 30.87603  & 18.463510 & 44.311332 &  0.000000 & 37.473682\\\\\n",
       "\tENST00000454820.5 &  0.00000  &  0.000000 &  0.000000 &  0.000000 &  0.000000\\\\\n",
       "\tENST00000535093.1 &  0.00000  &  0.000000 & 15.978063 &  0.000000 &  0.000000\\\\\n",
       "\tENST00000346219.7 &  0.00000  &  0.000000 &  7.487855 & 10.401091 &  8.667213\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | \n",
       "|---|---|---|---|---|\n",
       "| ENST00000548312.5 | 10.01273  |  8.991491 |  0.000000 |  8.237484 |  5.519675 | \n",
       "| ENST00000527779.1 | 30.87603  | 18.463510 | 44.311332 |  0.000000 | 37.473682 | \n",
       "| ENST00000454820.5 |  0.00000  |  0.000000 |  0.000000 |  0.000000 |  0.000000 | \n",
       "| ENST00000535093.1 |  0.00000  |  0.000000 | 15.978063 |  0.000000 |  0.000000 | \n",
       "| ENST00000346219.7 |  0.00000  |  0.000000 |  7.487855 | 10.401091 |  8.667213 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                  TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01\n",
       "ENST00000548312.5 10.01273         8.991491        0.000000      \n",
       "ENST00000527779.1 30.87603        18.463510       44.311332      \n",
       "ENST00000454820.5  0.00000         0.000000        0.000000      \n",
       "ENST00000535093.1  0.00000         0.000000       15.978063      \n",
       "ENST00000346219.7  0.00000         0.000000        7.487855      \n",
       "                  TCGA.A1.A0SM.01 TCGA.A1.A0SN.01\n",
       "ENST00000548312.5  8.237484        5.519675      \n",
       "ENST00000527779.1  0.000000       37.473682      \n",
       "ENST00000454820.5  0.000000        0.000000      \n",
       "ENST00000535093.1  0.000000        0.000000      \n",
       "ENST00000346219.7 10.401091        8.667213      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#examine transcript names in tx2gene\n",
    "head(tx2gene81)\n",
    "\n",
    "#examine transcript names in counts dataframe. Here we are looking at the first 5 rows and 5 columns\n",
    "brca.untran.counts.df[1:5, 1:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** the transcript names in the counts dataframe contains the version information. This is the number proceeding the period. tx2gene does not contain this information. We can tell tximport to ignore transcript versions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "summarizing abundance\n",
      "summarizing counts\n",
      "summarizing length\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "'list'"
      ],
      "text/latex": [
       "'list'"
      ],
      "text/markdown": [
       "'list'"
      ],
      "text/plain": [
       "[1] \"list\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#now to run tximport\n",
    "#we are only providing the counts and tpm data in the list, but we are inserting in dummy values for length as they are not used in this analysis\n",
    "#countsFromAbundance flag is set to no because these counts were generated directly from Kallisto and not backwards from tpm data\n",
    "#abundance is our untransformed transcript tpm data\n",
    "#counts is our untransformed transcript counts data\n",
    "txi <- list(abundance = brca.untran.tpm.df, counts = brca.untran.counts.df, length = brca.untran.counts.df, countsFromAbundance = \"no\")\n",
    "\n",
    "#txi is the list we made above with our transcript abundance and counts data\n",
    "#tx2gene81 is the annotation information to translate from transcript to gene\n",
    "#tx2gene81 was made in Notebook2\n",
    "\n",
    "df.txi <- summarizeToGene(txi, tx2gene81, ignoreTxVersion = TRUE)\n",
    "\n",
    "#what structure is df.txi\n",
    "str(df.txi)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 4\n",
      " $ abundance          :'data.frame':\t49922 obs. of  338 variables:\n",
      "  ..$ TCGA.3C.AALI.01: num [1:49922] 6.5153 0.0504 58.7344 33.5409 8.2979 ...\n",
      "  ..$ TCGA.3C.AALK.01: num [1:49922] 31.2981 0.0717 32.3299 10.595 5.8009 ...\n",
      "  ..$ TCGA.A1.A0SK.01: num [1:49922] 6.28e+01 -2.18e-08 7.86e+01 6.53 1.37e+01 ...\n",
      "  ..$ TCGA.A1.A0SM.01: num [1:49922] 31.5708 0.0154 46.3888 9.7262 7.9667 ...\n",
      "  ..$ TCGA.A1.A0SN.01: num [1:49922] 8.6704 0.0171 122.0497 14.3974 10.0153 ...\n",
      "  ..$ TCGA.A1.A0SO.01: num [1:49922] 47.3317 0.0989 93.0631 8.8083 16.1083 ...\n",
      "  ..$ TCGA.A1.A0SP.01: num [1:49922] 38.06 0.146 50.576 5.317 8.986 ...\n",
      "  ..$ TCGA.A2.A04P.01: num [1:49922] 25.9347 0.0346 13.4764 3.6072 3.3361 ...\n",
      "  ..$ TCGA.A2.A04Q.01: num [1:49922] 18.8883 0.0476 29.0192 8.5853 5.9036 ...\n",
      "  ..$ TCGA.A2.A04T.01: num [1:49922] 34.5 10.4 60.6 15.3 15.6 ...\n",
      "  ..$ TCGA.A2.A04W.01: num [1:49922] 30.4351 0.0195 48.5732 5.7227 3.833 ...\n",
      "  ..$ TCGA.A2.A04X.01: num [1:49922] 9.4439 0.0358 50.0049 5.1335 2.3086 ...\n",
      "  ..$ TCGA.A2.A0CM.01: num [1:49922] 26.009 0.045 84.329 8.058 10.994 ...\n",
      "  ..$ TCGA.A2.A0CX.01: num [1:49922] 30.0531 0.0165 41.1152 11.1658 7.9309 ...\n",
      "  ..$ TCGA.A2.A0D0.01: num [1:49922] 1.73e+01 -2.18e-08 8.42e+01 1.95 3.70 ...\n",
      "  ..$ TCGA.A2.A0D1.01: num [1:49922] 37.36 0.15 26.27 5.88 5.51 ...\n",
      "  ..$ TCGA.A2.A0D2.01: num [1:49922] 40.6916 0.0349 77.6579 2.8241 8.4348 ...\n",
      "  ..$ TCGA.A2.A0EY.01: num [1:49922] 19.528 0.388 47.848 13.214 7.758 ...\n",
      "  ..$ TCGA.A2.A0ST.01: num [1:49922] 26.591 0.417 32.937 5.936 5.122 ...\n",
      "  ..$ TCGA.A2.A0SX.01: num [1:49922] 18.215 0.524 64.968 15.4 14.694 ...\n",
      "  ..$ TCGA.A2.A0SY.01: num [1:49922] 16.39 1.26 33.45 10.38 4.87 ...\n",
      "  ..$ TCGA.A2.A0T0.01: num [1:49922] 45.2609 0.0141 59.0181 8.0538 12.7377 ...\n",
      "  ..$ TCGA.A2.A0T1.01: num [1:49922] 11.378 0.383 38.88 10.272 3.561 ...\n",
      "  ..$ TCGA.A2.A0T2.01: num [1:49922] 22.9116 0.0302 62.4068 7.5047 9.2053 ...\n",
      "  ..$ TCGA.A2.A0YE.01: num [1:49922] 107.406 0.142 77.196 8.744 12.01 ...\n",
      "  ..$ TCGA.A2.A0YG.01: num [1:49922] 12.1595 0.0577 83.6388 25.6624 24.6992 ...\n",
      "  ..$ TCGA.A2.A0YM.01: num [1:49922] 55.147 0.153 47.771 8.509 16.331 ...\n",
      "  ..$ TCGA.A2.A1G1.01: num [1:49922] 36.459 0.0126 89.2115 9.973 14.6049 ...\n",
      "  ..$ TCGA.A2.A1G6.01: num [1:49922] 58.15 2.08 60.03 11.44 3.6 ...\n",
      "  ..$ TCGA.A2.A3XS.01: num [1:49922] 70.722 0.229 68.314 8.212 6.591 ...\n",
      "  ..$ TCGA.A2.A3XT.01: num [1:49922] 9.398 0.313 68.867 9.241 13.587 ...\n",
      "  ..$ TCGA.A2.A3XU.01: num [1:49922] 17.5599 0.0215 15.4774 3.1641 4.6656 ...\n",
      "  ..$ TCGA.A2.A3XV.01: num [1:49922] 2.259 0.098 19.249 2.506 2 ...\n",
      "  ..$ TCGA.A2.A3XX.01: num [1:49922] 56.49 5.51 61.71 10.18 14.37 ...\n",
      "  ..$ TCGA.A2.A3XY.01: num [1:49922] 23 3.28 95.22 10.09 6.85 ...\n",
      "  ..$ TCGA.A2.A3XZ.01: num [1:49922] 24.702 0.191 36.842 23.486 4.73 ...\n",
      "  ..$ TCGA.A7.A0CE.01: num [1:49922] 77.07 1.05 49.64 9.02 18.9 ...\n",
      "  ..$ TCGA.A7.A0DA.01: num [1:49922] 26.108 0.118 42.803 12.722 8.84 ...\n",
      "  ..$ TCGA.A7.A26F.01: num [1:49922] 6.498 0.122 6.755 4.735 4.449 ...\n",
      "  ..$ TCGA.A7.A26G.01: num [1:49922] 71.1 1.24 71.54 13.84 11.06 ...\n",
      "  ..$ TCGA.A7.A26H.01: num [1:49922] 19.893 0.577 40.455 25.773 12.894 ...\n",
      "  ..$ TCGA.A7.A26I.01: num [1:49922] 31.4887 0.0434 48.2807 14.7767 46.1703 ...\n",
      "  ..$ TCGA.A7.A2KD.01: num [1:49922] 42.0509 0.0454 158.2 15.9196 11.3322 ...\n",
      "  ..$ TCGA.A7.A4SD.01: num [1:49922] 24.3254 0.0428 68.5575 7.3856 11.8467 ...\n",
      "  ..$ TCGA.A7.A4SE.01: num [1:49922] 35.2 1.08 80.18 4.97 13.97 ...\n",
      "  ..$ TCGA.A7.A4SF.01: num [1:49922] 4.68e+01 -2.18e-08 5.01e+01 1.11e+01 1.19e+01 ...\n",
      "  ..$ TCGA.A7.A5ZV.01: num [1:49922] 55.1731 0.0262 101.2002 8.0369 5.8263 ...\n",
      "  ..$ TCGA.A7.A6VV.01: num [1:49922] 23.87 0.194 59.632 4.546 4.174 ...\n",
      "  ..$ TCGA.A7.A6VW.01: num [1:49922] 39.9541 0.0464 30.5383 3.0742 11.4999 ...\n",
      "  ..$ TCGA.A7.A6VY.01: num [1:49922] 23.2409 0.0419 51.7092 7.5319 13.6934 ...\n",
      "  ..$ TCGA.A8.A06R.01: num [1:49922] 11.977 0.407 40.609 4.124 4.46 ...\n",
      "  ..$ TCGA.A8.A06T.01: num [1:49922] 50.7046 0.0423 54.5734 16.1781 5.2221 ...\n",
      "  ..$ TCGA.A8.A06U.01: num [1:49922] 2.38e+01 -2.18e-08 9.52e+01 8.76 6.87 ...\n",
      "  ..$ TCGA.A8.A06X.01: num [1:49922] 49.9007 0.0491 40.1211 13.9198 8.4933 ...\n",
      "  ..$ TCGA.A8.A075.01: num [1:49922] 10.647 0.203 48.103 6.894 5.437 ...\n",
      "  ..$ TCGA.A8.A076.01: num [1:49922] 16.449 0.682 86.78 15.952 8.798 ...\n",
      "  ..$ TCGA.A8.A07B.01: num [1:49922] 2.94e+01 -2.18e-08 4.27e+01 5.37 6.78 ...\n",
      "  ..$ TCGA.A8.A07C.01: num [1:49922] 38.3732 0.0528 112.4892 11.6235 22.2757 ...\n",
      "  ..$ TCGA.A8.A07I.01: num [1:49922] 15.188 0.921 45.126 10.385 8.087 ...\n",
      "  ..$ TCGA.A8.A07O.01: num [1:49922] 28.377 0.0131 48.3017 12.5941 20.3956 ...\n",
      "  ..$ TCGA.A8.A07P.01: num [1:49922] 11.99 0.364 56.401 11.497 6.96 ...\n",
      "  ..$ TCGA.A8.A07R.01: num [1:49922] 39.362 0.647 85.457 5.935 30.951 ...\n",
      "  ..$ TCGA.A8.A08B.01: num [1:49922] 1.58e+01 -2.18e-08 6.85e+01 1.25e+01 1.33e+01 ...\n",
      "  ..$ TCGA.A8.A08C.01: num [1:49922] 42.87 1.25 29.86 22.82 7.36 ...\n",
      "  ..$ TCGA.A8.A08G.01: num [1:49922] 13.275 0.169 82.573 22.927 10.967 ...\n",
      "  ..$ TCGA.A8.A08H.01: num [1:49922] 21.85 1.63 38.67 5.19 6.77 ...\n",
      "  ..$ TCGA.A8.A08P.01: num [1:49922] 10.899 0.125 95.246 10.476 10.364 ...\n",
      "  ..$ TCGA.A8.A08R.01: num [1:49922] 37.589 0.244 67.256 11.204 14.42 ...\n",
      "  ..$ TCGA.A8.A08S.01: num [1:49922] 9.5979 0.0631 45.842 12.7987 5.8982 ...\n",
      "  ..$ TCGA.A8.A08T.01: num [1:49922] 32.792 0.208 34.471 17.843 6.801 ...\n",
      "  ..$ TCGA.A8.A08X.01: num [1:49922] 1.15e+01 -2.18e-08 3.12e+01 4.76 5.28 ...\n",
      "  ..$ TCGA.A8.A090.01: num [1:49922] 17.341 0.253 69.519 15.911 11.058 ...\n",
      "  ..$ TCGA.A8.A097.01: num [1:49922] 26.586 0.197 67.423 15.647 9.87 ...\n",
      "  ..$ TCGA.A8.A099.01: num [1:49922] 36.4922 0.0177 61.9603 28.8646 10.5504 ...\n",
      "  ..$ TCGA.A8.A09E.01: num [1:49922] 42.21 0.134 61.891 13.788 11.2 ...\n",
      "  ..$ TCGA.A8.A09G.01: num [1:49922] 19.515 0.716 49.713 20.459 8.399 ...\n",
      "  ..$ TCGA.A8.A09I.01: num [1:49922] 7.893 0.124 76.147 22.004 10.509 ...\n",
      "  ..$ TCGA.A8.A09N.01: num [1:49922] 10.544 0.171 61.227 11.216 7.275 ...\n",
      "  ..$ TCGA.A8.A09X.01: num [1:49922] 12.587 0.569 20.541 8.823 6.824 ...\n",
      "  ..$ TCGA.A8.A0A7.01: num [1:49922] 9.888 0.396 56.169 7.902 6.455 ...\n",
      "  ..$ TCGA.A8.A0AB.01: num [1:49922] 141.506 0.548 35.306 26.549 24.545 ...\n",
      "  ..$ TCGA.AC.A23C.01: num [1:49922] 6.77 2.22 114.97 20.8 9.65 ...\n",
      "  ..$ TCGA.AC.A23G.01: num [1:49922] 26.997 0.242 27.989 12.788 7.614 ...\n",
      "  ..$ TCGA.AC.A23H.01: num [1:49922] 5.74 -2.18e-08 8.38e+01 3.21e+01 2.00e+01 ...\n",
      "  ..$ TCGA.AC.A2BK.01: num [1:49922] 9.30e+01 -2.18e-08 7.94e+01 3.07e+01 2.24e+01 ...\n",
      "  ..$ TCGA.AC.A2FB.01: num [1:49922] 39.157 0.293 32.284 13.288 7.407 ...\n",
      "  ..$ TCGA.AC.A2QH.01: num [1:49922] 8.98 49.87 7.37 1.33 2.22 ...\n",
      "  ..$ TCGA.AC.A2QJ.01: num [1:49922] 9.316 0.32 16.057 0.769 2.078 ...\n",
      "  ..$ TCGA.AC.A3QQ.01: num [1:49922] 8.756 0.706 7.241 7.37 6.919 ...\n",
      "  ..$ TCGA.AC.A3TN.01: num [1:49922] 15.63 0.625 34.66 10.77 3.775 ...\n",
      "  ..$ TCGA.AC.A3W5.01: num [1:49922] 30.01 0.678 36.793 14.351 4.799 ...\n",
      "  ..$ TCGA.AC.A3YI.01: num [1:49922] 23.653 0.115 45.592 9.945 4.235 ...\n",
      "  ..$ TCGA.AC.A4ZE.01: num [1:49922] 24.89 7.38 37.71 20.23 9.19 ...\n",
      "  ..$ TCGA.AC.A6IW.01: num [1:49922] 11.37 0.18 53.63 7.85 14.83 ...\n",
      "  ..$ TCGA.AC.A7VC.01: num [1:49922] 41.71 1.15 32.94 3.85 7.68 ...\n",
      "  ..$ TCGA.AC.A8OS.01: num [1:49922] 23.52 1.94 31.83 11.97 3.63 ...\n",
      "  ..$ TCGA.AN.A03X.01: num [1:49922] 14.7 2.2 46.1 10 8.9 ...\n",
      "  ..$ TCGA.AN.A041.01: num [1:49922] 23.6729 0.0838 36.6276 15.154 7.2657 ...\n",
      "  ..$ TCGA.AN.A04C.01: num [1:49922] 36.963 0.196 46.702 7.064 6.737 ...\n",
      "  .. [list output truncated]\n",
      " $ counts             :'data.frame':\t49922 obs. of  338 variables:\n",
      "  ..$ TCGA.3C.AALI.01: num [1:49922] 451 2 1821 3688 770 ...\n",
      "  ..$ TCGA.3C.AALK.01: num [1:49922] 3020 4 1405 1940 743 ...\n",
      "  ..$ TCGA.A1.A0SK.01: num [1:49922] 6486 0 3620 1324 1770 ...\n",
      "  ..$ TCGA.A1.A0SM.01: num [1:49922] 3642 1 2325 1972 1138 ...\n",
      "  ..$ TCGA.A1.A0SN.01: num [1:49922] 876 1 5509 2574 1300 ...\n",
      "  ..$ TCGA.A1.A0SO.01: num [1:49922] 5252 5 4678 1544 2157 ...\n",
      "  ..$ TCGA.A1.A0SP.01: num [1:49922] 4350 6 2524 1095 1349 ...\n",
      "  ..$ TCGA.A2.A04P.01: num [1:49922] 3965 3 874 1097 821 ...\n",
      "  ..$ TCGA.A2.A04Q.01: num [1:49922] 2835 4 1888 2412 1171 ...\n",
      "  ..$ TCGA.A2.A04T.01: num [1:49922] 4355 731 3316 3309 2411 ...\n",
      "  ..$ TCGA.A2.A04W.01: num [1:49922] 2779 1 1935 1089 504 ...\n",
      "  ..$ TCGA.A2.A04X.01: num [1:49922] 1430 3 3257 1337 508 ...\n",
      "  ..$ TCGA.A2.A0CM.01: num [1:49922] 2031 2 2911 1017 1100 ...\n",
      "  ..$ TCGA.A2.A0CX.01: num [1:49922] 3277 1 1924 2157 1158 ...\n",
      "  ..$ TCGA.A2.A0D0.01: num [1:49922] 2088 0 4437 540 629 ...\n",
      "  ..$ TCGA.A2.A0D1.01: num [1:49922] 3981 9 1219 1231 751 ...\n",
      "  ..$ TCGA.A2.A0D2.01: num [1:49922] 4128 2 3440 562 1084 ...\n",
      "  ..$ TCGA.A2.A0EY.01: num [1:49922] 2004 16 2143 2562 1021 ...\n",
      "  ..$ TCGA.A2.A0ST.01: num [1:49922] 3393 22 1809 1357 882 ...\n",
      "  ..$ TCGA.A2.A0SX.01: num [1:49922] 2215 29 3403 3249 2352 ...\n",
      "  ..$ TCGA.A2.A0SY.01: num [1:49922] 2056 70 1811 2315 851 ...\n",
      "  ..$ TCGA.A2.A0T0.01: num [1:49922] 5642 1 3242 1831 1969 ...\n",
      "  ..$ TCGA.A2.A0T1.01: num [1:49922] 1563 25 2345 2906 619 ...\n",
      "  ..$ TCGA.A2.A0T2.01: num [1:49922] 2700 2 3212 1721 1449 ...\n",
      "  ..$ TCGA.A2.A0YE.01: num [1:49922] 10568 8 3358 1467 1559 ...\n",
      "  ..$ TCGA.A2.A0YG.01: num [1:49922] 1113 3 3387 3650 3276 ...\n",
      "  ..$ TCGA.A2.A0YM.01: num [1:49922] 5536 9 2172 1394 1994 ...\n",
      "  ..$ TCGA.A2.A1G1.01: num [1:49922] 5220 1 5495 2299 2511 ...\n",
      "  ..$ TCGA.A2.A1G6.01: num [1:49922] 6374 88 2779 2234 525 ...\n",
      "  ..$ TCGA.A2.A3XS.01: num [1:49922] 9250 14 3902 1880 1123 ...\n",
      "  ..$ TCGA.A2.A3XT.01: num [1:49922] 1332 22 4255 2014 2478 ...\n",
      "  ..$ TCGA.A2.A3XU.01: num [1:49922] 2907 2 1101 984 1109 ...\n",
      "  ..$ TCGA.A2.A3XV.01: num [1:49922] 480 7 1630 1223 707 ...\n",
      "  ..$ TCGA.A2.A3XX.01: num [1:49922] 6908 377 3302 2095 2290 ...\n",
      "  ..$ TCGA.A2.A3XY.01: num [1:49922] 3237 218 5873 2302 1092 ...\n",
      "  ..$ TCGA.A2.A3XZ.01: num [1:49922] 2833 9 1952 4798 781 ...\n",
      "  ..$ TCGA.A7.A0CE.01: num [1:49922] 8487 54 2423 1694 2637 ...\n",
      "  ..$ TCGA.A7.A0DA.01: num [1:49922] 2771 7 1981 2432 1200 ...\n",
      "  ..$ TCGA.A7.A26F.01: num [1:49922] 2634.6 8.8 1368.5 2093.2 1236.4 ...\n",
      "  ..$ TCGA.A7.A26G.01: num [1:49922] 9920 69 4428 2998 2072 ...\n",
      "  ..$ TCGA.A7.A26H.01: num [1:49922] 2682 34 2464 5387 2325 ...\n",
      "  ..$ TCGA.A7.A26I.01: num [1:49922] 2183 9 1227 3004 5047 ...\n",
      "  ..$ TCGA.A7.A2KD.01: num [1:49922] 4995 3 8068 3233 1720 ...\n",
      "  ..$ TCGA.A7.A4SD.01: num [1:49922] 2032 2 2461 1012 1195 ...\n",
      "  ..$ TCGA.A7.A4SE.01: num [1:49922] 1915 30 1907 530 1034 ...\n",
      "  ..$ TCGA.A7.A4SF.01: num [1:49922] 3118 0 1436 1318 1041 ...\n",
      "  ..$ TCGA.A7.A5ZV.01: num [1:49922] 3786 1 2977 894 512 ...\n",
      "  ..$ TCGA.A7.A6VV.01: num [1:49922] 2244 5 2469 666 434 ...\n",
      "  ..$ TCGA.A7.A6VW.01: num [1:49922] 4694 3 1602 805 1800 ...\n",
      "  ..$ TCGA.A7.A6VY.01: num [1:49922] 2962 3 2886 1586 2216 ...\n",
      "  ..$ TCGA.A8.A06R.01: num [1:49922] 1359 9 2258 1091 677 ...\n",
      "  ..$ TCGA.A8.A06T.01: num [1:49922] 6100 3 3012 3570 888 ...\n",
      "  ..$ TCGA.A8.A06U.01: num [1:49922] 2461 0 4385 1504 995 ...\n",
      "  ..$ TCGA.A8.A06X.01: num [1:49922] 3396 2 1266 1507 882 ...\n",
      "  ..$ TCGA.A8.A075.01: num [1:49922] 1113 12 2230 1193 752 ...\n",
      "  ..$ TCGA.A8.A076.01: num [1:49922] 1809 30 4310 3007 1364 ...\n",
      "  ..$ TCGA.A8.A07B.01: num [1:49922] 1555 0 948 662 368 ...\n",
      "  ..$ TCGA.A8.A07C.01: num [1:49922] 5076 4 6605 2430 3961 ...\n",
      "  ..$ TCGA.A8.A07I.01: num [1:49922] 1172 28 1427 1364 701 ...\n",
      "  ..$ TCGA.A8.A07O.01: num [1:49922] 3835 1 2865 2674 3473 ...\n",
      "  ..$ TCGA.A8.A07P.01: num [1:49922] 1314 18 2754 2273 942 ...\n",
      "  ..$ TCGA.A8.A07R.01: num [1:49922] 5175 48 4886 1307 4629 ...\n",
      "  ..$ TCGA.A8.A08B.01: num [1:49922] 848 0 1389 1062 894 ...\n",
      "  ..$ TCGA.A8.A08C.01: num [1:49922] 4237 56 1280 3988 1004 ...\n",
      "  ..$ TCGA.A8.A08G.01: num [1:49922] 1517 11 4177 4732 1646 ...\n",
      "  ..$ TCGA.A8.A08H.01: num [1:49922] 1758 46 1359 785 643 ...\n",
      "  ..$ TCGA.A8.A08P.01: num [1:49922] 966 4 3660 1718 1224 ...\n",
      "  ..$ TCGA.A8.A08R.01: num [1:49922] 4534 17 3629 2113 2227 ...\n",
      "  ..$ TCGA.A8.A08S.01: num [1:49922] 1086 4 2260 2410 843 ...\n",
      "  ..$ TCGA.A8.A08T.01: num [1:49922] 3552 7 1653 3465 1010 ...\n",
      "  ..$ TCGA.A8.A08X.01: num [1:49922] 1572 0 1837 1191 934 ...\n",
      "  ..$ TCGA.A8.A090.01: num [1:49922] 1276 8 2205 2019 959 ...\n",
      "  ..$ TCGA.A8.A097.01: num [1:49922] 3513 12 3928 3316 1644 ...\n",
      "  ..$ TCGA.A8.A099.01: num [1:49922] 3558 1 2732 4401 1261 ...\n",
      "  ..$ TCGA.A8.A09E.01: num [1:49922] 3950 7 2508 2167 1212 ...\n",
      "  ..$ TCGA.A8.A09G.01: num [1:49922] 2438 33 2743 4511 1312 ...\n",
      "  ..$ TCGA.A8.A09I.01: num [1:49922] 1440 13 6170 6654 2568 ...\n",
      "  ..$ TCGA.A8.A09N.01: num [1:49922] 728 6 1861 1437 636 ...\n",
      "  ..$ TCGA.A8.A09X.01: num [1:49922] 1513 23 1027 1833 1113 ...\n",
      "  ..$ TCGA.A8.A0A7.01: num [1:49922] 1402 22 3563 1984 1075 ...\n",
      "  ..$ TCGA.A8.A0AB.01: num [1:49922] 16815 23 1813 4921 3608 ...\n",
      "  ..$ TCGA.AC.A23C.01: num [1:49922] 785 110 5952 4316 1492 ...\n",
      "  ..$ TCGA.AC.A23G.01: num [1:49922] 1376 7 625 1039 472 ...\n",
      "  ..$ TCGA.AC.A23H.01: num [1:49922] 643 0 4050 5852 3500 ...\n",
      "  ..$ TCGA.AC.A2BK.01: num [1:49922] 8623 0 3495 4883 2702 ...\n",
      "  ..$ TCGA.AC.A2FB.01: num [1:49922] 4625 11 1671 2861 1151 ...\n",
      "  ..$ TCGA.AC.A2QH.01: num [1:49922] 3823 11822 911 784 709 ...\n",
      "  ..$ TCGA.AC.A2QJ.01: num [1:49922] 1266 24 875 286 417 ...\n",
      "  ..$ TCGA.AC.A3QQ.01: num [1:49922] 2092.8 18.8 1001 2231 1611.4 ...\n",
      "  ..$ TCGA.AC.A3TN.01: num [1:49922] 2079 39 1943 2753 855 ...\n",
      "  ..$ TCGA.AC.A3W5.01: num [1:49922] 3146 26 1613 2861 692 ...\n",
      "  ..$ TCGA.AC.A3YI.01: num [1:49922] 1937 3 1625 1514 421 ...\n",
      "  ..$ TCGA.AC.A4ZE.01: num [1:49922] 2442 310 1673 3431 1194 ...\n",
      "  ..$ TCGA.AC.A6IW.01: num [1:49922] 1669 11 3512 2152 2812 ...\n",
      "  ..$ TCGA.AC.A7VC.01: num [1:49922] 3654 57 1278 652 883 ...\n",
      "  ..$ TCGA.AC.A8OS.01: num [1:49922] 2164 80 1325 1946 480 ...\n",
      "  ..$ TCGA.AN.A03X.01: num [1:49922] 1078 26 1395 1377 535 ...\n",
      "  ..$ TCGA.AN.A041.01: num [1:49922] 2476 3 1778 2870 1028 ...\n",
      "  ..$ TCGA.AN.A04C.01: num [1:49922] 6003 18 3325 2021 1597 ...\n",
      "  .. [list output truncated]\n",
      " $ length             :'data.frame':\t49922 obs. of  338 variables:\n",
      "  ..$ TCGA.3C.AALI.01: num [1:49922] 378 2 896 1889 185 ...\n",
      "  ..$ TCGA.3C.AALK.01: num [1:49922] 2603 4 910 623 154 ...\n",
      "  ..$ TCGA.A1.A0SK.01: num [1:49922] 4534 0 1666 504 375 ...\n",
      "  ..$ TCGA.A1.A0SM.01: num [1:49922] 3209 1 1900 841 376 ...\n",
      "  ..$ TCGA.A1.A0SN.01: num [1:49922] 709 1 3426 714 276 ...\n",
      "  ..$ TCGA.A1.A0SO.01: num [1:49922] 4805.24 3.21 2090.46 420.68 569.66 ...\n",
      "  ..$ TCGA.A1.A0SP.01: num [1:49922] 3780.91 2.93 2008.12 324.99 408.05 ...\n",
      "  ..$ TCGA.A2.A04P.01: num [1:49922] 3777 3 589 509 350 ...\n",
      "  ..$ TCGA.A2.A04Q.01: num [1:49922] 2559 4 1497 1070 436 ...\n",
      "  ..$ TCGA.A2.A04T.01: num [1:49922] 3777 731 2076 971 827 ...\n",
      "  ..$ TCGA.A2.A04W.01: num [1:49922] 2473 1 1325 499 152 ...\n",
      "  ..$ TCGA.A2.A04X.01: num [1:49922] 1261 3 2049 605 146 ...\n",
      "  ..$ TCGA.A2.A0CM.01: num [1:49922] 1870 2 2117 279 309 ...\n",
      "  ..$ TCGA.A2.A0CX.01: num [1:49922] 2909 1 1618 651 321 ...\n",
      "  ..$ TCGA.A2.A0D0.01: num [1:49922] 1920 0 2620 263 233 ...\n",
      "  ..$ TCGA.A2.A0D1.01: num [1:49922] 3619 9 924 524 283 ...\n",
      "  ..$ TCGA.A2.A0D2.01: num [1:49922] 3691 2 2496 198 326 ...\n",
      "  ..$ TCGA.A2.A0EY.01: num [1:49922] 1707.7 8.91 1584.36 821.89 231.74 ...\n",
      "  ..$ TCGA.A2.A0ST.01: num [1:49922] 2923 12.9 1455.3 438.2 283.6 ...\n",
      "  ..$ TCGA.A2.A0SX.01: num [1:49922] 1690.1 20.1 2079.8 982.1 575.8 ...\n",
      "  ..$ TCGA.A2.A0SY.01: num [1:49922] 1686.1 46.1 1199.3 783.3 258.2 ...\n",
      "  ..$ TCGA.A2.A0T0.01: num [1:49922] 4858 1 2412 713 737 ...\n",
      "  ..$ TCGA.A2.A0T1.01: num [1:49922] 1326.4 17.9 1931.2 1273.4 162 ...\n",
      "  ..$ TCGA.A2.A0T2.01: num [1:49922] 2542 2 2387 777 644 ...\n",
      "  ..$ TCGA.A2.A0YE.01: num [1:49922] 9564 8 2228 411 386 ...\n",
      "  ..$ TCGA.A2.A0YG.01: num [1:49922] 999 3 2810 917 1740 ...\n",
      "  ..$ TCGA.A2.A0YM.01: num [1:49922] 4629 9 1425 356 572 ...\n",
      "  ..$ TCGA.A2.A1G1.01: num [1:49922] 4773 1 3589 773 663 ...\n",
      "  ..$ TCGA.A2.A1G6.01: num [1:49922] 5275.8 48.5 2053.9 907.5 163.4 ...\n",
      "  ..$ TCGA.A2.A3XS.01: num [1:49922] 8049.83 9.99 2075.18 643.04 339.61 ...\n",
      "  ..$ TCGA.A2.A3XT.01: num [1:49922] 1164.3 17.3 2969 550.5 601.3 ...\n",
      "  ..$ TCGA.A2.A3XU.01: num [1:49922] 2463 2 625 429 500 ...\n",
      "  ..$ TCGA.A2.A3XV.01: num [1:49922] 401.25 3.32 997.86 703.17 653.6 ...\n",
      "  ..$ TCGA.A2.A3XX.01: num [1:49922] 6126 371 2337 599 783 ...\n",
      "  ..$ TCGA.A2.A3XY.01: num [1:49922] 2788 155 4859 626 320 ...\n",
      "  ..$ TCGA.A2.A3XZ.01: num [1:49922] 2206.7 4.8 1041.4 2070.9 230.7 ...\n",
      "  ..$ TCGA.A7.A0CE.01: num [1:49922] 7366 37 1445 424 991 ...\n",
      "  ..$ TCGA.A7.A0DA.01: num [1:49922] 2389 7 1123 1062 386 ...\n",
      "  ..$ TCGA.A7.A26F.01: num [1:49922] 2111.72 7.91 563.99 755.22 193.81 ...\n",
      "  ..$ TCGA.A7.A26G.01: num [1:49922] 9154.6 36.8 3688.8 773.9 462.6 ...\n",
      "  ..$ TCGA.A7.A26H.01: num [1:49922] 2190 20.1 1832.3 1455.1 648.7 ...\n",
      "  ..$ TCGA.A7.A26I.01: num [1:49922] 1838.67 4.36 897.5 833.99 1251.95 ...\n",
      "  ..$ TCGA.A7.A2KD.01: num [1:49922] 4377 3 6310 851 334 ...\n",
      "  ..$ TCGA.A7.A4SD.01: num [1:49922] 1857 2 1448 253 294 ...\n",
      "  ..$ TCGA.A7.A4SE.01: num [1:49922] 1613.6 24.3 1302.1 179.6 312.6 ...\n",
      "  ..$ TCGA.A7.A4SF.01: num [1:49922] 2664 0 1050 506 314 ...\n",
      "  ..$ TCGA.A7.A5ZV.01: num [1:49922] 3415 1 1776 246 190 ...\n",
      "  ..$ TCGA.A7.A6VV.01: num [1:49922] 1854.57 2.51 2042.57 212.12 104.91 ...\n",
      "  ..$ TCGA.A7.A6VW.01: num [1:49922] 4385.44 2.81 789.89 413.11 827.18 ...\n",
      "  ..$ TCGA.A7.A6VY.01: num [1:49922] 2552 3 1837 467 818 ...\n",
      "  ..$ TCGA.A8.A06R.01: num [1:49922] 972 9 1680 476 153 ...\n",
      "  ..$ TCGA.A8.A06T.01: num [1:49922] 4510 3 1597 1106 235 ...\n",
      "  ..$ TCGA.A8.A06U.01: num [1:49922] 2024 0 2401 407 276 ...\n",
      "  ..$ TCGA.A8.A06X.01: num [1:49922] 2841 2 820 434 280 ...\n",
      "  ..$ TCGA.A8.A075.01: num [1:49922] 988 12 1434 353 256 ...\n",
      "  ..$ TCGA.A8.A076.01: num [1:49922] 1584.8 15.8 2308.3 878.1 348.2 ...\n",
      "  ..$ TCGA.A8.A07B.01: num [1:49922] 1356.6 0 677.6 344 59.7 ...\n",
      "  ..$ TCGA.A8.A07C.01: num [1:49922] 4227 4 5789 750 1104 ...\n",
      "  ..$ TCGA.A8.A07I.01: num [1:49922] 991 15.6 853.4 443.2 160.9 ...\n",
      "  ..$ TCGA.A8.A07O.01: num [1:49922] 3502 1 2017 619 1005 ...\n",
      "  ..$ TCGA.A8.A07P.01: num [1:49922] 1022.2 11.6 1873.6 685 210.2 ...\n",
      "  ..$ TCGA.A8.A07R.01: num [1:49922] 4788 48 3610 378 1503 ...\n",
      "  ..$ TCGA.A8.A08B.01: num [1:49922] 699 0 931 349 280 ...\n",
      "  ..$ TCGA.A8.A08C.01: num [1:49922] 3571.4 37.8 957.6 1110.5 247.1 ...\n",
      "  ..$ TCGA.A8.A08G.01: num [1:49922] 1163 11 2393 1621 432 ...\n",
      "  ..$ TCGA.A8.A08H.01: num [1:49922] 1474 22 1063 250 213 ...\n",
      "  ..$ TCGA.A8.A08P.01: num [1:49922] 779.73 1.96 2421.91 748.73 418.31 ...\n",
      "  ..$ TCGA.A8.A08R.01: num [1:49922] 4076 17 2485 509 694 ...\n",
      "  ..$ TCGA.A8.A08S.01: num [1:49922] 865 4 1212 755 246 ...\n",
      "  ..$ TCGA.A8.A08T.01: num [1:49922] 3053.24 3.23 952.5 1031.2 274.62 ...\n",
      "  ..$ TCGA.A8.A08X.01: num [1:49922] 1311 0 1183 544 207 ...\n",
      "  ..$ TCGA.A8.A090.01: num [1:49922] 1030.97 5.02 1328.93 588.41 232.6 ...\n",
      "  ..$ TCGA.A8.A097.01: num [1:49922] 2955.21 8.12 2797.04 1021.61 432.52 ...\n",
      "  ..$ TCGA.A8.A099.01: num [1:49922] 3020 1 1553 1272 339 ...\n",
      "  ..$ TCGA.A8.A09E.01: num [1:49922] 3229 7 1686 624 349 ...\n",
      "  ..$ TCGA.A8.A09G.01: num [1:49922] 2159 16.3 2094.2 1425.9 290.4 ...\n",
      "  ..$ TCGA.A8.A09I.01: num [1:49922] 1202 13 4773 2928 653 ...\n",
      "  ..$ TCGA.A8.A09N.01: num [1:49922] 506.98 4.85 1219.36 429.3 172.29 ...\n",
      "  ..$ TCGA.A8.A09X.01: num [1:49922] 1281.5 10.9 676 554.4 300.6 ...\n",
      "  ..$ TCGA.A8.A0A7.01: num [1:49922] 1221.3 11.4 2462.8 561 313.7 ...\n",
      "  ..$ TCGA.A8.A0AB.01: num [1:49922] 14599.2 11.1 1211.9 1168.5 920.6 ...\n",
      "  ..$ TCGA.AC.A23C.01: num [1:49922] 603.1 64.5 4552.6 1719.6 372.9 ...\n",
      "  ..$ TCGA.AC.A23G.01: num [1:49922] 1173 7 323 265 154 ...\n",
      "  ..$ TCGA.AC.A23H.01: num [1:49922] 545 0 2868 1907 2683 ...\n",
      "  ..$ TCGA.AC.A2BK.01: num [1:49922] 6172 0 2551 1757 585 ...\n",
      "  ..$ TCGA.AC.A2FB.01: num [1:49922] 4001.94 5.06 1260.44 897.22 237.94 ...\n",
      "  ..$ TCGA.AC.A2QH.01: num [1:49922] 2851 11698 275 259 189 ...\n",
      "  ..$ TCGA.AC.A2QJ.01: num [1:49922] 1099 24 359 241 228 ...\n",
      "  ..$ TCGA.AC.A3QQ.01: num [1:49922] 1461.9 11.2 580.3 655.1 223.2 ...\n",
      "  ..$ TCGA.AC.A3TN.01: num [1:49922] 1797.8 28.5 1212.3 1306 540.6 ...\n",
      "  ..$ TCGA.AC.A3W5.01: num [1:49922] 2549 13.5 1097.1 1215.9 290.5 ...\n",
      "  ..$ TCGA.AC.A3YI.01: num [1:49922] 1604.76 1.38 1109.75 565.46 124.97 ...\n",
      "  ..$ TCGA.AC.A4ZE.01: num [1:49922] 1934 178 1118 981 289 ...\n",
      "  ..$ TCGA.AC.A6IW.01: num [1:49922] 1462.64 6.15 3041.3 687.26 1272.09 ...\n",
      "  ..$ TCGA.AC.A7VC.01: num [1:49922] 3324 57 765 214 377 ...\n",
      "  ..$ TCGA.AC.A8OS.01: num [1:49922] 1693 49 862 562 82 ...\n",
      "  ..$ TCGA.AN.A03X.01: num [1:49922] 844 26 841 588 157 ...\n",
      "  ..$ TCGA.AN.A041.01: num [1:49922] 1960.74 1.38 855.65 756.03 203.6 ...\n",
      "  ..$ TCGA.AN.A04C.01: num [1:49922] 5644 18 2546 1004 514 ...\n",
      "  .. [list output truncated]\n",
      " $ countsFromAbundance: chr \"no\"\n"
     ]
    }
   ],
   "source": [
    "str(df.txi)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "as we can see with the str(df.txi) call, df.txi is a list of all gene-level counts and abundance data for **49922 genes**. Now we need to extract the gene level counts and tpm data from df.txi. We will do this and coerce each into a dataframe for later use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "counts data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A1.A0SO.01</th><th scope=col>TCGA.A1.A0SP.01</th><th scope=col>TCGA.A2.A04P.01</th><th scope=col>TCGA.A2.A04Q.01</th><th scope=col>TCGA.A2.A04T.01</th><th scope=col>⋯</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.PE.A5DC.01</th><th scope=col>TCGA.PE.A5DD.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA14.01</th><th scope=col>TCGA.S3.AA15.01</th><th scope=col>TCGA.UL.AAZ6.01</th><th scope=col>TCGA.UU.A93S.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td> 450.655142</td><td>3020.109054</td><td>6486.0687  </td><td>3641.6943  </td><td> 875.8434  </td><td>5252.389068</td><td>4349.910381</td><td>3964.5744  </td><td>2835.217596</td><td>4355.0541  </td><td>⋯          </td><td> 645.52538 </td><td>4536.742724</td><td>3691.4786  </td><td>2400.522368</td><td>1371.52306 </td><td>1661.2877  </td><td>2090.7229  </td><td>1186.40144 </td><td> 976.8824  </td><td> 364.1602  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td>   2.000078</td><td>   3.999903</td><td>   0.0000  </td><td>   1.0000  </td><td>   1.0000  </td><td>   4.999912</td><td>   6.000185</td><td>   3.0000  </td><td>   3.999903</td><td> 731.0001  </td><td>⋯          </td><td>  11.00031 </td><td>   6.000219</td><td>3556.0146  </td><td>   5.000156</td><td>  24.99974 </td><td>   0.0000  </td><td>   7.0000  </td><td>  37.99979 </td><td>   1.0000  </td><td>  36.0012  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td>1821.031845</td><td>1405.018826</td><td>3620.0691  </td><td>2324.9518  </td><td>5508.9806  </td><td>4677.870576</td><td>2524.012514</td><td> 873.9736  </td><td>1888.007985</td><td>3316.0325  </td><td>⋯          </td><td>1597.98718 </td><td>2047.962746</td><td>1358.0296  </td><td>1512.008500</td><td>1189.99584 </td><td>2402.0106  </td><td>1015.0289  </td><td>1221.00364 </td><td>1163.9743  </td><td>1328.0154  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td>3688.499306</td><td>1940.060272</td><td>1323.6190  </td><td>1972.4032  </td><td>2573.8937  </td><td>1543.572332</td><td>1095.212755</td><td>1096.8912  </td><td>2412.455125</td><td>3308.7035  </td><td>⋯          </td><td> 710.14563 </td><td>1259.116073</td><td> 754.7742  </td><td>1227.154984</td><td>2042.30419 </td><td> 549.9404  </td><td>1087.0107  </td><td> 961.86779 </td><td>1708.3689  </td><td>1660.7872  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td> 769.629818</td><td> 743.029409</td><td>1769.5213  </td><td>1138.3500  </td><td>1300.2310  </td><td>2156.624673</td><td>1349.009842</td><td> 820.7999  </td><td>1170.767933</td><td>2410.5476  </td><td>⋯          </td><td> 528.78813 </td><td>1924.489990</td><td> 364.9363  </td><td>1044.840011</td><td>1096.15716 </td><td> 725.6080  </td><td> 509.8673  </td><td>1310.12223 </td><td> 856.8481  </td><td> 957.7056  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td> 383.997866</td><td> 385.992250</td><td> 115.9989  </td><td> 689.0194  </td><td> 411.9968  </td><td> 191.001862</td><td> 924.994900</td><td> 381.0095  </td><td>1708.974047</td><td> 839.0132  </td><td>⋯          </td><td> 685.98680 </td><td> 461.996608</td><td> 185.0003  </td><td> 310.004006</td><td> 478.99150 </td><td>1183.9836  </td><td> 169.0000  </td><td> 955.01035 </td><td> 215.9941  </td><td> 186.0013  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A1.A0SO.01 & TCGA.A1.A0SP.01 & TCGA.A2.A04P.01 & TCGA.A2.A04Q.01 & TCGA.A2.A04T.01 & ⋯ & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.PE.A5DC.01 & TCGA.PE.A5DD.01 & TCGA.S3.AA10.01 & TCGA.S3.AA14.01 & TCGA.S3.AA15.01 & TCGA.UL.AAZ6.01 & TCGA.UU.A93S.01\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &  450.655142 & 3020.109054 & 6486.0687   & 3641.6943   &  875.8434   & 5252.389068 & 4349.910381 & 3964.5744   & 2835.217596 & 4355.0541   & ⋯           &  645.52538  & 4536.742724 & 3691.4786   & 2400.522368 & 1371.52306  & 1661.2877   & 2090.7229   & 1186.40144  &  976.8824   &  364.1602  \\\\\n",
       "\tENSG00000000005 &    2.000078 &    3.999903 &    0.0000   &    1.0000   &    1.0000   &    4.999912 &    6.000185 &    3.0000   &    3.999903 &  731.0001   & ⋯           &   11.00031  &    6.000219 & 3556.0146   &    5.000156 &   24.99974  &    0.0000   &    7.0000   &   37.99979  &    1.0000   &   36.0012  \\\\\n",
       "\tENSG00000000419 & 1821.031845 & 1405.018826 & 3620.0691   & 2324.9518   & 5508.9806   & 4677.870576 & 2524.012514 &  873.9736   & 1888.007985 & 3316.0325   & ⋯           & 1597.98718  & 2047.962746 & 1358.0296   & 1512.008500 & 1189.99584  & 2402.0106   & 1015.0289   & 1221.00364  & 1163.9743   & 1328.0154  \\\\\n",
       "\tENSG00000000457 & 3688.499306 & 1940.060272 & 1323.6190   & 1972.4032   & 2573.8937   & 1543.572332 & 1095.212755 & 1096.8912   & 2412.455125 & 3308.7035   & ⋯           &  710.14563  & 1259.116073 &  754.7742   & 1227.154984 & 2042.30419  &  549.9404   & 1087.0107   &  961.86779  & 1708.3689   & 1660.7872  \\\\\n",
       "\tENSG00000000460 &  769.629818 &  743.029409 & 1769.5213   & 1138.3500   & 1300.2310   & 2156.624673 & 1349.009842 &  820.7999   & 1170.767933 & 2410.5476   & ⋯           &  528.78813  & 1924.489990 &  364.9363   & 1044.840011 & 1096.15716  &  725.6080   &  509.8673   & 1310.12223  &  856.8481   &  957.7056  \\\\\n",
       "\tENSG00000000938 &  383.997866 &  385.992250 &  115.9989   &  689.0194   &  411.9968   &  191.001862 &  924.994900 &  381.0095   & 1708.974047 &  839.0132   & ⋯           &  685.98680  &  461.996608 &  185.0003   &  310.004006 &  478.99150  & 1183.9836   &  169.0000   &  955.01035  &  215.9941   &  186.0013  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A1.A0SO.01 | TCGA.A1.A0SP.01 | TCGA.A2.A04P.01 | TCGA.A2.A04Q.01 | TCGA.A2.A04T.01 | ⋯ | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.PE.A5DC.01 | TCGA.PE.A5DD.01 | TCGA.S3.AA10.01 | TCGA.S3.AA14.01 | TCGA.S3.AA15.01 | TCGA.UL.AAZ6.01 | TCGA.UU.A93S.01 | \n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |  450.655142 | 3020.109054 | 6486.0687   | 3641.6943   |  875.8434   | 5252.389068 | 4349.910381 | 3964.5744   | 2835.217596 | 4355.0541   | ⋯           |  645.52538  | 4536.742724 | 3691.4786   | 2400.522368 | 1371.52306  | 1661.2877   | 2090.7229   | 1186.40144  |  976.8824   |  364.1602   | \n",
       "| ENSG00000000005 |    2.000078 |    3.999903 |    0.0000   |    1.0000   |    1.0000   |    4.999912 |    6.000185 |    3.0000   |    3.999903 |  731.0001   | ⋯           |   11.00031  |    6.000219 | 3556.0146   |    5.000156 |   24.99974  |    0.0000   |    7.0000   |   37.99979  |    1.0000   |   36.0012   | \n",
       "| ENSG00000000419 | 1821.031845 | 1405.018826 | 3620.0691   | 2324.9518   | 5508.9806   | 4677.870576 | 2524.012514 |  873.9736   | 1888.007985 | 3316.0325   | ⋯           | 1597.98718  | 2047.962746 | 1358.0296   | 1512.008500 | 1189.99584  | 2402.0106   | 1015.0289   | 1221.00364  | 1163.9743   | 1328.0154   | \n",
       "| ENSG00000000457 | 3688.499306 | 1940.060272 | 1323.6190   | 1972.4032   | 2573.8937   | 1543.572332 | 1095.212755 | 1096.8912   | 2412.455125 | 3308.7035   | ⋯           |  710.14563  | 1259.116073 |  754.7742   | 1227.154984 | 2042.30419  |  549.9404   | 1087.0107   |  961.86779  | 1708.3689   | 1660.7872   | \n",
       "| ENSG00000000460 |  769.629818 |  743.029409 | 1769.5213   | 1138.3500   | 1300.2310   | 2156.624673 | 1349.009842 |  820.7999   | 1170.767933 | 2410.5476   | ⋯           |  528.78813  | 1924.489990 |  364.9363   | 1044.840011 | 1096.15716  |  725.6080   |  509.8673   | 1310.12223  |  856.8481   |  957.7056   | \n",
       "| ENSG00000000938 |  383.997866 |  385.992250 |  115.9989   |  689.0194   |  411.9968   |  191.001862 |  924.994900 |  381.0095   | 1708.974047 |  839.0132   | ⋯           |  685.98680  |  461.996608 |  185.0003   |  310.004006 |  478.99150  | 1183.9836   |  169.0000   |  955.01035  |  215.9941   |  186.0013   | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01 TCGA.A1.A0SM.01\n",
       "ENSG00000000003  450.655142     3020.109054     6486.0687       3641.6943      \n",
       "ENSG00000000005    2.000078        3.999903        0.0000          1.0000      \n",
       "ENSG00000000419 1821.031845     1405.018826     3620.0691       2324.9518      \n",
       "ENSG00000000457 3688.499306     1940.060272     1323.6190       1972.4032      \n",
       "ENSG00000000460  769.629818      743.029409     1769.5213       1138.3500      \n",
       "ENSG00000000938  383.997866      385.992250      115.9989        689.0194      \n",
       "                TCGA.A1.A0SN.01 TCGA.A1.A0SO.01 TCGA.A1.A0SP.01 TCGA.A2.A04P.01\n",
       "ENSG00000000003  875.8434       5252.389068     4349.910381     3964.5744      \n",
       "ENSG00000000005    1.0000          4.999912        6.000185        3.0000      \n",
       "ENSG00000000419 5508.9806       4677.870576     2524.012514      873.9736      \n",
       "ENSG00000000457 2573.8937       1543.572332     1095.212755     1096.8912      \n",
       "ENSG00000000460 1300.2310       2156.624673     1349.009842      820.7999      \n",
       "ENSG00000000938  411.9968        191.001862      924.994900      381.0095      \n",
       "                TCGA.A2.A04Q.01 TCGA.A2.A04T.01 ⋯ TCGA.OL.A66P.01\n",
       "ENSG00000000003 2835.217596     4355.0541       ⋯  645.52538     \n",
       "ENSG00000000005    3.999903      731.0001       ⋯   11.00031     \n",
       "ENSG00000000419 1888.007985     3316.0325       ⋯ 1597.98718     \n",
       "ENSG00000000457 2412.455125     3308.7035       ⋯  710.14563     \n",
       "ENSG00000000460 1170.767933     2410.5476       ⋯  528.78813     \n",
       "ENSG00000000938 1708.974047      839.0132       ⋯  685.98680     \n",
       "                TCGA.OL.A6VO.01 TCGA.OL.A97C.01 TCGA.PE.A5DC.01 TCGA.PE.A5DD.01\n",
       "ENSG00000000003 4536.742724     3691.4786       2400.522368     1371.52306     \n",
       "ENSG00000000005    6.000219     3556.0146          5.000156       24.99974     \n",
       "ENSG00000000419 2047.962746     1358.0296       1512.008500     1189.99584     \n",
       "ENSG00000000457 1259.116073      754.7742       1227.154984     2042.30419     \n",
       "ENSG00000000460 1924.489990      364.9363       1044.840011     1096.15716     \n",
       "ENSG00000000938  461.996608      185.0003        310.004006      478.99150     \n",
       "                TCGA.S3.AA10.01 TCGA.S3.AA14.01 TCGA.S3.AA15.01 TCGA.UL.AAZ6.01\n",
       "ENSG00000000003 1661.2877       2090.7229       1186.40144       976.8824      \n",
       "ENSG00000000005    0.0000          7.0000         37.99979         1.0000      \n",
       "ENSG00000000419 2402.0106       1015.0289       1221.00364      1163.9743      \n",
       "ENSG00000000457  549.9404       1087.0107        961.86779      1708.3689      \n",
       "ENSG00000000460  725.6080        509.8673       1310.12223       856.8481      \n",
       "ENSG00000000938 1183.9836        169.0000        955.01035       215.9941      \n",
       "                TCGA.UU.A93S.01\n",
       "ENSG00000000003  364.1602      \n",
       "ENSG00000000005   36.0012      \n",
       "ENSG00000000419 1328.0154      \n",
       "ENSG00000000457 1660.7872      \n",
       "ENSG00000000460  957.7056      \n",
       "ENSG00000000938  186.0013      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tpm data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A1.A0SO.01</th><th scope=col>TCGA.A1.A0SP.01</th><th scope=col>TCGA.A2.A04P.01</th><th scope=col>TCGA.A2.A04Q.01</th><th scope=col>TCGA.A2.A04T.01</th><th scope=col>⋯</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.PE.A5DC.01</th><th scope=col>TCGA.PE.A5DD.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA14.01</th><th scope=col>TCGA.S3.AA15.01</th><th scope=col>TCGA.UL.AAZ6.01</th><th scope=col>TCGA.UU.A93S.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td> 6.51534946  </td><td>31.29805585  </td><td> 6.280911e+01</td><td>31.57077798  </td><td>  8.67041684 </td><td>47.33168141  </td><td>38.059551    </td><td>25.93473872  </td><td>18.88830498  </td><td>34.544688    </td><td>⋯            </td><td> 4.4470885   </td><td>45.5182870   </td><td> 67.884085   </td><td>22.16971860  </td><td>17.3165612   </td><td> 1.651789e+01</td><td>31.8898304   </td><td> 9.242428    </td><td> 7.05971952  </td><td> 4.9771771   </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td> 0.05044589  </td><td> 0.07170003  </td><td>-2.178597e-08</td><td> 0.01536661  </td><td>  0.01712719 </td><td> 0.09889547  </td><td> 0.146012    </td><td> 0.03461922  </td><td> 0.04761009  </td><td>10.390269    </td><td>⋯            </td><td> 0.1356351   </td><td> 0.1028231   </td><td>115.969091   </td><td> 0.08478087  </td><td> 0.7063652   </td><td>-2.178597e-08</td><td> 0.1873907   </td><td> 0.689528    </td><td> 0.01311623  </td><td> 0.8968211   </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td>58.73442028  </td><td>32.32993541  </td><td> 7.859218e+01</td><td>46.38880189  </td><td>122.04966013 </td><td>93.06309574  </td><td>50.576155    </td><td>13.47643623  </td><td>29.01918536  </td><td>60.612913    </td><td>⋯            </td><td>25.2861948   </td><td>45.2074857   </td><td> 56.744003   </td><td>33.11791951  </td><td>33.2652626   </td><td> 5.324058e+01</td><td>34.9271880   </td><td>21.119278    </td><td>19.93375070  </td><td>42.4709247   </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td>33.54091741  </td><td>10.59498195  </td><td> 6.533141e+00</td><td> 9.72619167  </td><td> 14.39741632 </td><td> 8.80833987  </td><td> 5.317215    </td><td> 3.60721561  </td><td> 8.58530095  </td><td>15.318031    </td><td>⋯            </td><td> 2.7659811   </td><td> 6.9569049   </td><td>  8.621612   </td><td> 6.30240497  </td><td>15.0684562   </td><td> 3.374201e+00</td><td> 9.6647475   </td><td> 4.437816    </td><td> 6.65371733  </td><td>11.1486465   </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td> 8.29792941  </td><td> 5.80092100  </td><td> 1.373277e+01</td><td> 7.96667122  </td><td> 10.01529016 </td><td>16.10825740  </td><td> 8.985589    </td><td> 3.33609228  </td><td> 5.90359760  </td><td>15.559711    </td><td>⋯            </td><td> 2.1640055   </td><td>15.1271171   </td><td>  5.323229   </td><td> 6.50888160  </td><td> 9.9350641   </td><td> 5.685957e+00</td><td> 5.9254496   </td><td> 7.828398    </td><td> 4.01198994  </td><td>10.2266756   </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td> 5.89488620  </td><td> 4.21810256  </td><td> 1.056610e+00</td><td> 6.45857008  </td><td>  4.07883811 </td><td> 1.68323530  </td><td> 7.893180    </td><td> 2.47017262  </td><td>11.13667583  </td><td> 6.771309    </td><td>⋯            </td><td> 4.4954092   </td><td> 4.3927861   </td><td>  3.793558   </td><td> 3.13872136  </td><td> 5.7674670   </td><td> 1.177756e+01</td><td> 2.7647520   </td><td> 7.713530    </td><td> 1.69585923  </td><td> 2.5440392   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A1.A0SO.01 & TCGA.A1.A0SP.01 & TCGA.A2.A04P.01 & TCGA.A2.A04Q.01 & TCGA.A2.A04T.01 & ⋯ & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.PE.A5DC.01 & TCGA.PE.A5DD.01 & TCGA.S3.AA10.01 & TCGA.S3.AA14.01 & TCGA.S3.AA15.01 & TCGA.UL.AAZ6.01 & TCGA.UU.A93S.01\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &  6.51534946   & 31.29805585   &  6.280911e+01 & 31.57077798   &   8.67041684  & 47.33168141   & 38.059551     & 25.93473872   & 18.88830498   & 34.544688     & ⋯             &  4.4470885    & 45.5182870    &  67.884085    & 22.16971860   & 17.3165612    &  1.651789e+01 & 31.8898304    &  9.242428     &  7.05971952   &  4.9771771   \\\\\n",
       "\tENSG00000000005 &  0.05044589   &  0.07170003   & -2.178597e-08 &  0.01536661   &   0.01712719  &  0.09889547   &  0.146012     &  0.03461922   &  0.04761009   & 10.390269     & ⋯             &  0.1356351    &  0.1028231    & 115.969091    &  0.08478087   &  0.7063652    & -2.178597e-08 &  0.1873907    &  0.689528     &  0.01311623   &  0.8968211   \\\\\n",
       "\tENSG00000000419 & 58.73442028   & 32.32993541   &  7.859218e+01 & 46.38880189   & 122.04966013  & 93.06309574   & 50.576155     & 13.47643623   & 29.01918536   & 60.612913     & ⋯             & 25.2861948    & 45.2074857    &  56.744003    & 33.11791951   & 33.2652626    &  5.324058e+01 & 34.9271880    & 21.119278     & 19.93375070   & 42.4709247   \\\\\n",
       "\tENSG00000000457 & 33.54091741   & 10.59498195   &  6.533141e+00 &  9.72619167   &  14.39741632  &  8.80833987   &  5.317215     &  3.60721561   &  8.58530095   & 15.318031     & ⋯             &  2.7659811    &  6.9569049    &   8.621612    &  6.30240497   & 15.0684562    &  3.374201e+00 &  9.6647475    &  4.437816     &  6.65371733   & 11.1486465   \\\\\n",
       "\tENSG00000000460 &  8.29792941   &  5.80092100   &  1.373277e+01 &  7.96667122   &  10.01529016  & 16.10825740   &  8.985589     &  3.33609228   &  5.90359760   & 15.559711     & ⋯             &  2.1640055    & 15.1271171    &   5.323229    &  6.50888160   &  9.9350641    &  5.685957e+00 &  5.9254496    &  7.828398     &  4.01198994   & 10.2266756   \\\\\n",
       "\tENSG00000000938 &  5.89488620   &  4.21810256   &  1.056610e+00 &  6.45857008   &   4.07883811  &  1.68323530   &  7.893180     &  2.47017262   & 11.13667583   &  6.771309     & ⋯             &  4.4954092    &  4.3927861    &   3.793558    &  3.13872136   &  5.7674670    &  1.177756e+01 &  2.7647520    &  7.713530     &  1.69585923   &  2.5440392   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A1.A0SO.01 | TCGA.A1.A0SP.01 | TCGA.A2.A04P.01 | TCGA.A2.A04Q.01 | TCGA.A2.A04T.01 | ⋯ | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.PE.A5DC.01 | TCGA.PE.A5DD.01 | TCGA.S3.AA10.01 | TCGA.S3.AA14.01 | TCGA.S3.AA15.01 | TCGA.UL.AAZ6.01 | TCGA.UU.A93S.01 | \n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |  6.51534946   | 31.29805585   |  6.280911e+01 | 31.57077798   |   8.67041684  | 47.33168141   | 38.059551     | 25.93473872   | 18.88830498   | 34.544688     | ⋯             |  4.4470885    | 45.5182870    |  67.884085    | 22.16971860   | 17.3165612    |  1.651789e+01 | 31.8898304    |  9.242428     |  7.05971952   |  4.9771771    | \n",
       "| ENSG00000000005 |  0.05044589   |  0.07170003   | -2.178597e-08 |  0.01536661   |   0.01712719  |  0.09889547   |  0.146012     |  0.03461922   |  0.04761009   | 10.390269     | ⋯             |  0.1356351    |  0.1028231    | 115.969091    |  0.08478087   |  0.7063652    | -2.178597e-08 |  0.1873907    |  0.689528     |  0.01311623   |  0.8968211    | \n",
       "| ENSG00000000419 | 58.73442028   | 32.32993541   |  7.859218e+01 | 46.38880189   | 122.04966013  | 93.06309574   | 50.576155     | 13.47643623   | 29.01918536   | 60.612913     | ⋯             | 25.2861948    | 45.2074857    |  56.744003    | 33.11791951   | 33.2652626    |  5.324058e+01 | 34.9271880    | 21.119278     | 19.93375070   | 42.4709247    | \n",
       "| ENSG00000000457 | 33.54091741   | 10.59498195   |  6.533141e+00 |  9.72619167   |  14.39741632  |  8.80833987   |  5.317215     |  3.60721561   |  8.58530095   | 15.318031     | ⋯             |  2.7659811    |  6.9569049    |   8.621612    |  6.30240497   | 15.0684562    |  3.374201e+00 |  9.6647475    |  4.437816     |  6.65371733   | 11.1486465    | \n",
       "| ENSG00000000460 |  8.29792941   |  5.80092100   |  1.373277e+01 |  7.96667122   |  10.01529016  | 16.10825740   |  8.985589     |  3.33609228   |  5.90359760   | 15.559711     | ⋯             |  2.1640055    | 15.1271171    |   5.323229    |  6.50888160   |  9.9350641    |  5.685957e+00 |  5.9254496    |  7.828398     |  4.01198994   | 10.2266756    | \n",
       "| ENSG00000000938 |  5.89488620   |  4.21810256   |  1.056610e+00 |  6.45857008   |   4.07883811  |  1.68323530   |  7.893180     |  2.47017262   | 11.13667583   |  6.771309     | ⋯             |  4.4954092    |  4.3927861    |   3.793558    |  3.13872136   |  5.7674670    |  1.177756e+01 |  2.7647520    |  7.713530     |  1.69585923   |  2.5440392    | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01 TCGA.A1.A0SM.01\n",
       "ENSG00000000003  6.51534946     31.29805585      6.280911e+01   31.57077798    \n",
       "ENSG00000000005  0.05044589      0.07170003     -2.178597e-08    0.01536661    \n",
       "ENSG00000000419 58.73442028     32.32993541      7.859218e+01   46.38880189    \n",
       "ENSG00000000457 33.54091741     10.59498195      6.533141e+00    9.72619167    \n",
       "ENSG00000000460  8.29792941      5.80092100      1.373277e+01    7.96667122    \n",
       "ENSG00000000938  5.89488620      4.21810256      1.056610e+00    6.45857008    \n",
       "                TCGA.A1.A0SN.01 TCGA.A1.A0SO.01 TCGA.A1.A0SP.01 TCGA.A2.A04P.01\n",
       "ENSG00000000003   8.67041684    47.33168141     38.059551       25.93473872    \n",
       "ENSG00000000005   0.01712719     0.09889547      0.146012        0.03461922    \n",
       "ENSG00000000419 122.04966013    93.06309574     50.576155       13.47643623    \n",
       "ENSG00000000457  14.39741632     8.80833987      5.317215        3.60721561    \n",
       "ENSG00000000460  10.01529016    16.10825740      8.985589        3.33609228    \n",
       "ENSG00000000938   4.07883811     1.68323530      7.893180        2.47017262    \n",
       "                TCGA.A2.A04Q.01 TCGA.A2.A04T.01 ⋯ TCGA.OL.A66P.01\n",
       "ENSG00000000003 18.88830498     34.544688       ⋯  4.4470885     \n",
       "ENSG00000000005  0.04761009     10.390269       ⋯  0.1356351     \n",
       "ENSG00000000419 29.01918536     60.612913       ⋯ 25.2861948     \n",
       "ENSG00000000457  8.58530095     15.318031       ⋯  2.7659811     \n",
       "ENSG00000000460  5.90359760     15.559711       ⋯  2.1640055     \n",
       "ENSG00000000938 11.13667583      6.771309       ⋯  4.4954092     \n",
       "                TCGA.OL.A6VO.01 TCGA.OL.A97C.01 TCGA.PE.A5DC.01 TCGA.PE.A5DD.01\n",
       "ENSG00000000003 45.5182870       67.884085      22.16971860     17.3165612     \n",
       "ENSG00000000005  0.1028231      115.969091       0.08478087      0.7063652     \n",
       "ENSG00000000419 45.2074857       56.744003      33.11791951     33.2652626     \n",
       "ENSG00000000457  6.9569049        8.621612       6.30240497     15.0684562     \n",
       "ENSG00000000460 15.1271171        5.323229       6.50888160      9.9350641     \n",
       "ENSG00000000938  4.3927861        3.793558       3.13872136      5.7674670     \n",
       "                TCGA.S3.AA10.01 TCGA.S3.AA14.01 TCGA.S3.AA15.01 TCGA.UL.AAZ6.01\n",
       "ENSG00000000003  1.651789e+01   31.8898304       9.242428        7.05971952    \n",
       "ENSG00000000005 -2.178597e-08    0.1873907       0.689528        0.01311623    \n",
       "ENSG00000000419  5.324058e+01   34.9271880      21.119278       19.93375070    \n",
       "ENSG00000000457  3.374201e+00    9.6647475       4.437816        6.65371733    \n",
       "ENSG00000000460  5.685957e+00    5.9254496       7.828398        4.01198994    \n",
       "ENSG00000000938  1.177756e+01    2.7647520       7.713530        1.69585923    \n",
       "                TCGA.UU.A93S.01\n",
       "ENSG00000000003  4.9771771     \n",
       "ENSG00000000005  0.8968211     \n",
       "ENSG00000000419 42.4709247     \n",
       "ENSG00000000457 11.1486465     \n",
       "ENSG00000000460 10.2266756     \n",
       "ENSG00000000938  2.5440392     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#after tximport we need to create variables with the gene counts and tpm\n",
    "brca.untran.gene.counts <- as.data.frame(df.txi$counts)\n",
    "brca.untran.gene.tpm <- as.data.frame(df.txi$abundance)\n",
    "\n",
    "\n",
    "#to view the gene counts data. note that the rownames start with ENSG instead of ENST\n",
    "message(\"counts data\")\n",
    "head(brca.untran.gene.counts, nrows = 10)\n",
    "\n",
    "message(\"tpm data\")\n",
    "head(brca.untran.gene.tpm, nrows = 10)\n",
    "\n",
    "#save the data\n",
    "save(brca.untran.gene.counts, file = file.path(date, \"brca.untran.gene.counts.Rda\"))\n",
    "save(brca.untran.gene.tpm, file = file.path(date, \"brca.untran.gene.tpm.Rda\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** the tpm data is lower than the counts data for each gene because it is a normalized value correcting for gene length and sequencing depth. [This](https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/) blog does a nice job explaining how the calculation is performed and compares it to two other common RNAseq normalization methods rpkm and fpkm.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Check the results of tximport\n",
    "Let's also look at the difference between the output of tximport and the input counts file.\n",
    "First let's identify the transcripts from the first gene listed in the counts file above, \"ENSG00000000003\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'ENST00000612152'</li>\n",
       "\t<li>'ENST00000373020'</li>\n",
       "\t<li>'ENST00000614008'</li>\n",
       "\t<li>'ENST00000496771'</li>\n",
       "\t<li>'ENST00000494424'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ENST00000612152'\n",
       "\\item 'ENST00000373020'\n",
       "\\item 'ENST00000614008'\n",
       "\\item 'ENST00000496771'\n",
       "\\item 'ENST00000494424'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ENST00000612152'\n",
       "2. 'ENST00000373020'\n",
       "3. 'ENST00000614008'\n",
       "4. 'ENST00000496771'\n",
       "5. 'ENST00000494424'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"ENST00000612152\" \"ENST00000373020\" \"ENST00000614008\" \"ENST00000496771\"\n",
       "[5] \"ENST00000494424\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene1_transcripts <- as.vector(tx2gene81[tx2gene81$gene == \"ENSG00000000003\", 1])\n",
    "gene1_transcripts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the difference between the output and input of tximport in relation to 'ENSG00000000003'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gene level counts\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A1.A0SO.01</th><th scope=col>TCGA.A1.A0SP.01</th><th scope=col>TCGA.A2.A04P.01</th><th scope=col>TCGA.A2.A04Q.01</th><th scope=col>TCGA.A2.A04T.01</th><th scope=col>⋯</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.PE.A5DC.01</th><th scope=col>TCGA.PE.A5DD.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA14.01</th><th scope=col>TCGA.S3.AA15.01</th><th scope=col>TCGA.UL.AAZ6.01</th><th scope=col>TCGA.UU.A93S.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td>450.6551</td><td>3020.109</td><td>6486.069</td><td>3641.694</td><td>875.8434</td><td>5252.389</td><td>4349.91 </td><td>3964.574</td><td>2835.218</td><td>4355.054</td><td>⋯       </td><td>645.5254</td><td>4536.743</td><td>3691.479</td><td>2400.522</td><td>1371.523</td><td>1661.288</td><td>2090.723</td><td>1186.401</td><td>976.8824</td><td>364.1602</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A1.A0SO.01 & TCGA.A1.A0SP.01 & TCGA.A2.A04P.01 & TCGA.A2.A04Q.01 & TCGA.A2.A04T.01 & ⋯ & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.PE.A5DC.01 & TCGA.PE.A5DD.01 & TCGA.S3.AA10.01 & TCGA.S3.AA14.01 & TCGA.S3.AA15.01 & TCGA.UL.AAZ6.01 & TCGA.UU.A93S.01\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 & 450.6551 & 3020.109 & 6486.069 & 3641.694 & 875.8434 & 5252.389 & 4349.91  & 3964.574 & 2835.218 & 4355.054 & ⋯        & 645.5254 & 4536.743 & 3691.479 & 2400.522 & 1371.523 & 1661.288 & 2090.723 & 1186.401 & 976.8824 & 364.1602\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A1.A0SO.01 | TCGA.A1.A0SP.01 | TCGA.A2.A04P.01 | TCGA.A2.A04Q.01 | TCGA.A2.A04T.01 | ⋯ | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.PE.A5DC.01 | TCGA.PE.A5DD.01 | TCGA.S3.AA10.01 | TCGA.S3.AA14.01 | TCGA.S3.AA15.01 | TCGA.UL.AAZ6.01 | TCGA.UU.A93S.01 | \n",
       "|---|\n",
       "| ENSG00000000003 | 450.6551 | 3020.109 | 6486.069 | 3641.694 | 875.8434 | 5252.389 | 4349.91  | 3964.574 | 2835.218 | 4355.054 | ⋯        | 645.5254 | 4536.743 | 3691.479 | 2400.522 | 1371.523 | 1661.288 | 2090.723 | 1186.401 | 976.8824 | 364.1602 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01 TCGA.A1.A0SM.01\n",
       "ENSG00000000003 450.6551        3020.109        6486.069        3641.694       \n",
       "                TCGA.A1.A0SN.01 TCGA.A1.A0SO.01 TCGA.A1.A0SP.01 TCGA.A2.A04P.01\n",
       "ENSG00000000003 875.8434        5252.389        4349.91         3964.574       \n",
       "                TCGA.A2.A04Q.01 TCGA.A2.A04T.01 ⋯ TCGA.OL.A66P.01\n",
       "ENSG00000000003 2835.218        4355.054        ⋯ 645.5254       \n",
       "                TCGA.OL.A6VO.01 TCGA.OL.A97C.01 TCGA.PE.A5DC.01 TCGA.PE.A5DD.01\n",
       "ENSG00000000003 4536.743        3691.479        2400.522        1371.523       \n",
       "                TCGA.S3.AA10.01 TCGA.S3.AA14.01 TCGA.S3.AA15.01 TCGA.UL.AAZ6.01\n",
       "ENSG00000000003 1661.288        2090.723        1186.401        976.8824       \n",
       "                TCGA.UU.A93S.01\n",
       "ENSG00000000003 364.1602       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "transcript level counts\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A1.A0SO.01</th><th scope=col>TCGA.A1.A0SP.01</th><th scope=col>TCGA.A2.A04P.01</th><th scope=col>TCGA.A2.A04Q.01</th><th scope=col>TCGA.A2.A04T.01</th><th scope=col>⋯</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.PE.A5DC.01</th><th scope=col>TCGA.PE.A5DD.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA14.01</th><th scope=col>TCGA.S3.AA15.01</th><th scope=col>TCGA.UL.AAZ6.01</th><th scope=col>TCGA.UU.A93S.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENST00000612152.4</th><td> 30.825251 </td><td> 161.60187 </td><td> 752.205644</td><td> 198.68747 </td><td> 57.258635 </td><td> 128.723856</td><td> 270.291493</td><td> 116.474189</td><td> 158.421235</td><td> 324.60973 </td><td>⋯          </td><td> 49.202942 </td><td> 133.093919</td><td> 284.13474 </td><td> 320.840037</td><td>  77.20560 </td><td> 119.01742 </td><td> 216.89662 </td><td>  51.95151 </td><td>127.417677 </td><td> 40.24936  </td></tr>\n",
       "\t<tr><th scope=row>ENST00000373020.8</th><td>411.028408 </td><td>2805.97403 </td><td>5422.468177</td><td>3403.90776 </td><td>792.727511 </td><td>5047.363228</td><td>4028.541566</td><td>3845.418434</td><td>2665.996100</td><td>4000.70749 </td><td>⋯          </td><td>594.558741 </td><td>4306.992432</td><td>3369.62403 </td><td>2050.125426</td><td>1238.29162 </td><td>1506.45581 </td><td>1820.61202 </td><td>1113.36026 </td><td>840.138756 </td><td>323.91082  </td></tr>\n",
       "\t<tr><th scope=row>ENST00000614008.4</th><td>  0.000000 </td><td>   0.00000 </td><td>   9.074246</td><td>  17.51086 </td><td>  6.543322 </td><td>   9.526107</td><td>   4.208977</td><td>   0.000000</td><td>   5.293227</td><td>   0.00000 </td><td>⋯          </td><td>  0.000000 </td><td>   8.411339</td><td>   0.00000 </td><td>   5.367822</td><td>   0.00000 </td><td>   0.00000 </td><td>   0.00000 </td><td>   0.00000 </td><td>  0.000000 </td><td>  0.00000  </td></tr>\n",
       "\t<tr><th scope=row>ENST00000496771.5</th><td>  8.801483 </td><td>  38.16790 </td><td> 302.320669</td><td>  21.58824 </td><td> 19.313966 </td><td>  66.775878</td><td>  46.868344</td><td>   2.681771</td><td>   5.507034</td><td>  29.73691 </td><td>⋯          </td><td>  1.763698 </td><td>  88.245035</td><td>  37.71981 </td><td>  22.231332</td><td>  49.54863 </td><td>  35.81444 </td><td>  53.21424 </td><td>  21.08967 </td><td>  9.325931 </td><td>  0.00000  </td></tr>\n",
       "\t<tr><th scope=row>ENST00000494424.1</th><td>  0.000000 </td><td>  14.36526 </td><td>   0.000000</td><td>   0.00000 </td><td>  0.000000 </td><td>   0.000000</td><td>   0.000000</td><td>   0.000000</td><td>   0.000000</td><td>   0.00000 </td><td>⋯          </td><td>  0.000000 </td><td>   0.000000</td><td>   0.00000 </td><td>   1.957750</td><td>   6.47721 </td><td>   0.00000 </td><td>   0.00000 </td><td>   0.00000 </td><td>  0.000000 </td><td>  0.00000  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A1.A0SO.01 & TCGA.A1.A0SP.01 & TCGA.A2.A04P.01 & TCGA.A2.A04Q.01 & TCGA.A2.A04T.01 & ⋯ & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.PE.A5DC.01 & TCGA.PE.A5DD.01 & TCGA.S3.AA10.01 & TCGA.S3.AA14.01 & TCGA.S3.AA15.01 & TCGA.UL.AAZ6.01 & TCGA.UU.A93S.01\\\\\n",
       "\\hline\n",
       "\tENST00000612152.4 &  30.825251  &  161.60187  &  752.205644 &  198.68747  &  57.258635  &  128.723856 &  270.291493 &  116.474189 &  158.421235 &  324.60973  & ⋯           &  49.202942  &  133.093919 &  284.13474  &  320.840037 &   77.20560  &  119.01742  &  216.89662  &   51.95151  & 127.417677  &  40.24936  \\\\\n",
       "\tENST00000373020.8 & 411.028408  & 2805.97403  & 5422.468177 & 3403.90776  & 792.727511  & 5047.363228 & 4028.541566 & 3845.418434 & 2665.996100 & 4000.70749  & ⋯           & 594.558741  & 4306.992432 & 3369.62403  & 2050.125426 & 1238.29162  & 1506.45581  & 1820.61202  & 1113.36026  & 840.138756  & 323.91082  \\\\\n",
       "\tENST00000614008.4 &   0.000000  &    0.00000  &    9.074246 &   17.51086  &   6.543322  &    9.526107 &    4.208977 &    0.000000 &    5.293227 &    0.00000  & ⋯           &   0.000000  &    8.411339 &    0.00000  &    5.367822 &    0.00000  &    0.00000  &    0.00000  &    0.00000  &   0.000000  &   0.00000  \\\\\n",
       "\tENST00000496771.5 &   8.801483  &   38.16790  &  302.320669 &   21.58824  &  19.313966  &   66.775878 &   46.868344 &    2.681771 &    5.507034 &   29.73691  & ⋯           &   1.763698  &   88.245035 &   37.71981  &   22.231332 &   49.54863  &   35.81444  &   53.21424  &   21.08967  &   9.325931  &   0.00000  \\\\\n",
       "\tENST00000494424.1 &   0.000000  &   14.36526  &    0.000000 &    0.00000  &   0.000000  &    0.000000 &    0.000000 &    0.000000 &    0.000000 &    0.00000  & ⋯           &   0.000000  &    0.000000 &    0.00000  &    1.957750 &    6.47721  &    0.00000  &    0.00000  &    0.00000  &   0.000000  &   0.00000  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A1.A0SO.01 | TCGA.A1.A0SP.01 | TCGA.A2.A04P.01 | TCGA.A2.A04Q.01 | TCGA.A2.A04T.01 | ⋯ | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.PE.A5DC.01 | TCGA.PE.A5DD.01 | TCGA.S3.AA10.01 | TCGA.S3.AA14.01 | TCGA.S3.AA15.01 | TCGA.UL.AAZ6.01 | TCGA.UU.A93S.01 | \n",
       "|---|---|---|---|---|\n",
       "| ENST00000612152.4 |  30.825251  |  161.60187  |  752.205644 |  198.68747  |  57.258635  |  128.723856 |  270.291493 |  116.474189 |  158.421235 |  324.60973  | ⋯           |  49.202942  |  133.093919 |  284.13474  |  320.840037 |   77.20560  |  119.01742  |  216.89662  |   51.95151  | 127.417677  |  40.24936   | \n",
       "| ENST00000373020.8 | 411.028408  | 2805.97403  | 5422.468177 | 3403.90776  | 792.727511  | 5047.363228 | 4028.541566 | 3845.418434 | 2665.996100 | 4000.70749  | ⋯           | 594.558741  | 4306.992432 | 3369.62403  | 2050.125426 | 1238.29162  | 1506.45581  | 1820.61202  | 1113.36026  | 840.138756  | 323.91082   | \n",
       "| ENST00000614008.4 |   0.000000  |    0.00000  |    9.074246 |   17.51086  |   6.543322  |    9.526107 |    4.208977 |    0.000000 |    5.293227 |    0.00000  | ⋯           |   0.000000  |    8.411339 |    0.00000  |    5.367822 |    0.00000  |    0.00000  |    0.00000  |    0.00000  |   0.000000  |   0.00000   | \n",
       "| ENST00000496771.5 |   8.801483  |   38.16790  |  302.320669 |   21.58824  |  19.313966  |   66.775878 |   46.868344 |    2.681771 |    5.507034 |   29.73691  | ⋯           |   1.763698  |   88.245035 |   37.71981  |   22.231332 |   49.54863  |   35.81444  |   53.21424  |   21.08967  |   9.325931  |   0.00000   | \n",
       "| ENST00000494424.1 |   0.000000  |   14.36526  |    0.000000 |    0.00000  |   0.000000  |    0.000000 |    0.000000 |    0.000000 |    0.000000 |    0.00000  | ⋯           |   0.000000  |    0.000000 |    0.00000  |    1.957750 |    6.47721  |    0.00000  |    0.00000  |    0.00000  |   0.000000  |   0.00000   | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                  TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01\n",
       "ENST00000612152.4  30.825251       161.60187       752.205644    \n",
       "ENST00000373020.8 411.028408      2805.97403      5422.468177    \n",
       "ENST00000614008.4   0.000000         0.00000         9.074246    \n",
       "ENST00000496771.5   8.801483        38.16790       302.320669    \n",
       "ENST00000494424.1   0.000000        14.36526         0.000000    \n",
       "                  TCGA.A1.A0SM.01 TCGA.A1.A0SN.01 TCGA.A1.A0SO.01\n",
       "ENST00000612152.4  198.68747       57.258635       128.723856    \n",
       "ENST00000373020.8 3403.90776      792.727511      5047.363228    \n",
       "ENST00000614008.4   17.51086        6.543322         9.526107    \n",
       "ENST00000496771.5   21.58824       19.313966        66.775878    \n",
       "ENST00000494424.1    0.00000        0.000000         0.000000    \n",
       "                  TCGA.A1.A0SP.01 TCGA.A2.A04P.01 TCGA.A2.A04Q.01\n",
       "ENST00000612152.4  270.291493      116.474189      158.421235    \n",
       "ENST00000373020.8 4028.541566     3845.418434     2665.996100    \n",
       "ENST00000614008.4    4.208977        0.000000        5.293227    \n",
       "ENST00000496771.5   46.868344        2.681771        5.507034    \n",
       "ENST00000494424.1    0.000000        0.000000        0.000000    \n",
       "                  TCGA.A2.A04T.01 ⋯ TCGA.OL.A66P.01 TCGA.OL.A6VO.01\n",
       "ENST00000612152.4  324.60973      ⋯  49.202942       133.093919    \n",
       "ENST00000373020.8 4000.70749      ⋯ 594.558741      4306.992432    \n",
       "ENST00000614008.4    0.00000      ⋯   0.000000         8.411339    \n",
       "ENST00000496771.5   29.73691      ⋯   1.763698        88.245035    \n",
       "ENST00000494424.1    0.00000      ⋯   0.000000         0.000000    \n",
       "                  TCGA.OL.A97C.01 TCGA.PE.A5DC.01 TCGA.PE.A5DD.01\n",
       "ENST00000612152.4  284.13474       320.840037       77.20560     \n",
       "ENST00000373020.8 3369.62403      2050.125426     1238.29162     \n",
       "ENST00000614008.4    0.00000         5.367822        0.00000     \n",
       "ENST00000496771.5   37.71981        22.231332       49.54863     \n",
       "ENST00000494424.1    0.00000         1.957750        6.47721     \n",
       "                  TCGA.S3.AA10.01 TCGA.S3.AA14.01 TCGA.S3.AA15.01\n",
       "ENST00000612152.4  119.01742       216.89662        51.95151     \n",
       "ENST00000373020.8 1506.45581      1820.61202      1113.36026     \n",
       "ENST00000614008.4    0.00000         0.00000         0.00000     \n",
       "ENST00000496771.5   35.81444        53.21424        21.08967     \n",
       "ENST00000494424.1    0.00000         0.00000         0.00000     \n",
       "                  TCGA.UL.AAZ6.01 TCGA.UU.A93S.01\n",
       "ENST00000612152.4 127.417677       40.24936      \n",
       "ENST00000373020.8 840.138756      323.91082      \n",
       "ENST00000614008.4   0.000000        0.00000      \n",
       "ENST00000496771.5   9.325931        0.00000      \n",
       "ENST00000494424.1   0.000000        0.00000      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "message(\"gene level counts\")\n",
    "brca.untran.gene.counts[\"ENSG00000000003\", ]\n",
    "\n",
    "message(\"transcript level counts\")\n",
    "brca.untran.counts.df[gene1_transcripts, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that the resulting gene level count data is essentially the sum of all the transcripts. This confirms that tximport worked (at least for this gene)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Calculate differential expressed genes using [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)\n",
    "DESeq2 is a powerful tool that performs differential gene expression analysis based on the negative binomial distribution, using normalized count data. Again, I highly recommend reading the [manual](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) before performing the next steps. There is even a [coursera lecture discussing DESeq2](https://www.coursera.org/lecture/bioconductor/count-based-rna-seq-analysis-XVfPX) that you might want to watch. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run DESeq2 we need to specify:\n",
    "1. A dataframe explaining all samples (TCGA IDs) and their perscribed \"group.\" For ex. TCGA.XX.XXXY.01 is TNBC and TCGA.XX.XXXX.01 is HER2.\n",
    "2. The counts data with TCGA identifiers that are exactly the same as the dataframe above.\n",
    "3. The comparison. This is easy for this example because there are only 2 groups and we are calculating the DE genes between those groups.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Before creating the three items needed, let's check that the TCGA identifiers are exactly the same between the counts and clinical data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>sampleID</th><th scope=col>breast_carcinoma_estrogen_receptor_status</th><th scope=col>breast_carcinoma_progesterone_receptor_status</th><th scope=col>lab_proc_her2_neu_immunohistochemistry_receptor_status</th><th scope=col>lab_procedure_her2_neu_in_situ_hybrid_outcome_type</th><th scope=col>HER2</th><th scope=col>TNBC</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>TCGA-3C-AAAU-01</td><td>Positive       </td><td>Positive       </td><td>Negative       </td><td>unknown        </td><td>Negative       </td><td>NOT_TNBC       </td></tr>\n",
       "\t<tr><td>TCGA-3C-AALI-01</td><td>Positive       </td><td>Positive       </td><td>Positive       </td><td>unknown        </td><td>Positive       </td><td>NOT_TNBC       </td></tr>\n",
       "\t<tr><td>TCGA-3C-AALJ-01</td><td>Positive       </td><td>Positive       </td><td>Indeterminate  </td><td>unknown        </td><td>unknown        </td><td>NOT_TNBC       </td></tr>\n",
       "\t<tr><td>TCGA-3C-AALK-01</td><td>Positive       </td><td>Positive       </td><td>Positive       </td><td>unknown        </td><td>Positive       </td><td>NOT_TNBC       </td></tr>\n",
       "\t<tr><td>TCGA-4H-AAAK-01</td><td>Positive       </td><td>Positive       </td><td>Equivocal      </td><td>unknown        </td><td>unknown        </td><td>NOT_TNBC       </td></tr>\n",
       "\t<tr><td>TCGA-5L-AAT0-01</td><td>Positive       </td><td>Positive       </td><td>Negative       </td><td>unknown        </td><td>Negative       </td><td>NOT_TNBC       </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllll}\n",
       " sampleID & breast\\_carcinoma\\_estrogen\\_receptor\\_status & breast\\_carcinoma\\_progesterone\\_receptor\\_status & lab\\_proc\\_her2\\_neu\\_immunohistochemistry\\_receptor\\_status & lab\\_procedure\\_her2\\_neu\\_in\\_situ\\_hybrid\\_outcome\\_type & HER2 & TNBC\\\\\n",
       "\\hline\n",
       "\t TCGA-3C-AAAU-01 & Positive        & Positive        & Negative        & unknown         & Negative        & NOT\\_TNBC     \\\\\n",
       "\t TCGA-3C-AALI-01 & Positive        & Positive        & Positive        & unknown         & Positive        & NOT\\_TNBC     \\\\\n",
       "\t TCGA-3C-AALJ-01 & Positive        & Positive        & Indeterminate   & unknown         & unknown         & NOT\\_TNBC     \\\\\n",
       "\t TCGA-3C-AALK-01 & Positive        & Positive        & Positive        & unknown         & Positive        & NOT\\_TNBC     \\\\\n",
       "\t TCGA-4H-AAAK-01 & Positive        & Positive        & Equivocal       & unknown         & unknown         & NOT\\_TNBC     \\\\\n",
       "\t TCGA-5L-AAT0-01 & Positive        & Positive        & Negative        & unknown         & Negative        & NOT\\_TNBC     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "sampleID | breast_carcinoma_estrogen_receptor_status | breast_carcinoma_progesterone_receptor_status | lab_proc_her2_neu_immunohistochemistry_receptor_status | lab_procedure_her2_neu_in_situ_hybrid_outcome_type | HER2 | TNBC | \n",
       "|---|---|---|---|---|---|\n",
       "| TCGA-3C-AAAU-01 | Positive        | Positive        | Negative        | unknown         | Negative        | NOT_TNBC        | \n",
       "| TCGA-3C-AALI-01 | Positive        | Positive        | Positive        | unknown         | Positive        | NOT_TNBC        | \n",
       "| TCGA-3C-AALJ-01 | Positive        | Positive        | Indeterminate   | unknown         | unknown         | NOT_TNBC        | \n",
       "| TCGA-3C-AALK-01 | Positive        | Positive        | Positive        | unknown         | Positive        | NOT_TNBC        | \n",
       "| TCGA-4H-AAAK-01 | Positive        | Positive        | Equivocal       | unknown         | unknown         | NOT_TNBC        | \n",
       "| TCGA-5L-AAT0-01 | Positive        | Positive        | Negative        | unknown         | Negative        | NOT_TNBC        | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  sampleID        breast_carcinoma_estrogen_receptor_status\n",
       "1 TCGA-3C-AAAU-01 Positive                                 \n",
       "2 TCGA-3C-AALI-01 Positive                                 \n",
       "3 TCGA-3C-AALJ-01 Positive                                 \n",
       "4 TCGA-3C-AALK-01 Positive                                 \n",
       "5 TCGA-4H-AAAK-01 Positive                                 \n",
       "6 TCGA-5L-AAT0-01 Positive                                 \n",
       "  breast_carcinoma_progesterone_receptor_status\n",
       "1 Positive                                     \n",
       "2 Positive                                     \n",
       "3 Positive                                     \n",
       "4 Positive                                     \n",
       "5 Positive                                     \n",
       "6 Positive                                     \n",
       "  lab_proc_her2_neu_immunohistochemistry_receptor_status\n",
       "1 Negative                                              \n",
       "2 Positive                                              \n",
       "3 Indeterminate                                         \n",
       "4 Positive                                              \n",
       "5 Equivocal                                             \n",
       "6 Negative                                              \n",
       "  lab_procedure_her2_neu_in_situ_hybrid_outcome_type HER2     TNBC    \n",
       "1 unknown                                            Negative NOT_TNBC\n",
       "2 unknown                                            Positive NOT_TNBC\n",
       "3 unknown                                            unknown  NOT_TNBC\n",
       "4 unknown                                            Positive NOT_TNBC\n",
       "5 unknown                                            unknown  NOT_TNBC\n",
       "6 unknown                                            Negative NOT_TNBC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A1.A0SO.01</th><th scope=col>TCGA.A1.A0SP.01</th><th scope=col>TCGA.A2.A04P.01</th><th scope=col>TCGA.A2.A04Q.01</th><th scope=col>TCGA.A2.A04T.01</th><th scope=col>⋯</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.PE.A5DC.01</th><th scope=col>TCGA.PE.A5DD.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA14.01</th><th scope=col>TCGA.S3.AA15.01</th><th scope=col>TCGA.UL.AAZ6.01</th><th scope=col>TCGA.UU.A93S.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td> 450.655142</td><td>3020.109054</td><td>6486.0687  </td><td>3641.6943  </td><td> 875.8434  </td><td>5252.389068</td><td>4349.910381</td><td>3964.5744  </td><td>2835.217596</td><td>4355.0541  </td><td>⋯          </td><td> 645.52538 </td><td>4536.742724</td><td>3691.4786  </td><td>2400.522368</td><td>1371.52306 </td><td>1661.2877  </td><td>2090.7229  </td><td>1186.40144 </td><td> 976.8824  </td><td> 364.1602  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td>   2.000078</td><td>   3.999903</td><td>   0.0000  </td><td>   1.0000  </td><td>   1.0000  </td><td>   4.999912</td><td>   6.000185</td><td>   3.0000  </td><td>   3.999903</td><td> 731.0001  </td><td>⋯          </td><td>  11.00031 </td><td>   6.000219</td><td>3556.0146  </td><td>   5.000156</td><td>  24.99974 </td><td>   0.0000  </td><td>   7.0000  </td><td>  37.99979 </td><td>   1.0000  </td><td>  36.0012  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td>1821.031845</td><td>1405.018826</td><td>3620.0691  </td><td>2324.9518  </td><td>5508.9806  </td><td>4677.870576</td><td>2524.012514</td><td> 873.9736  </td><td>1888.007985</td><td>3316.0325  </td><td>⋯          </td><td>1597.98718 </td><td>2047.962746</td><td>1358.0296  </td><td>1512.008500</td><td>1189.99584 </td><td>2402.0106  </td><td>1015.0289  </td><td>1221.00364 </td><td>1163.9743  </td><td>1328.0154  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td>3688.499306</td><td>1940.060272</td><td>1323.6190  </td><td>1972.4032  </td><td>2573.8937  </td><td>1543.572332</td><td>1095.212755</td><td>1096.8912  </td><td>2412.455125</td><td>3308.7035  </td><td>⋯          </td><td> 710.14563 </td><td>1259.116073</td><td> 754.7742  </td><td>1227.154984</td><td>2042.30419 </td><td> 549.9404  </td><td>1087.0107  </td><td> 961.86779 </td><td>1708.3689  </td><td>1660.7872  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td> 769.629818</td><td> 743.029409</td><td>1769.5213  </td><td>1138.3500  </td><td>1300.2310  </td><td>2156.624673</td><td>1349.009842</td><td> 820.7999  </td><td>1170.767933</td><td>2410.5476  </td><td>⋯          </td><td> 528.78813 </td><td>1924.489990</td><td> 364.9363  </td><td>1044.840011</td><td>1096.15716 </td><td> 725.6080  </td><td> 509.8673  </td><td>1310.12223 </td><td> 856.8481  </td><td> 957.7056  </td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td> 383.997866</td><td> 385.992250</td><td> 115.9989  </td><td> 689.0194  </td><td> 411.9968  </td><td> 191.001862</td><td> 924.994900</td><td> 381.0095  </td><td>1708.974047</td><td> 839.0132  </td><td>⋯          </td><td> 685.98680 </td><td> 461.996608</td><td> 185.0003  </td><td> 310.004006</td><td> 478.99150 </td><td>1183.9836  </td><td> 169.0000  </td><td> 955.01035 </td><td> 215.9941  </td><td> 186.0013  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A1.A0SO.01 & TCGA.A1.A0SP.01 & TCGA.A2.A04P.01 & TCGA.A2.A04Q.01 & TCGA.A2.A04T.01 & ⋯ & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.PE.A5DC.01 & TCGA.PE.A5DD.01 & TCGA.S3.AA10.01 & TCGA.S3.AA14.01 & TCGA.S3.AA15.01 & TCGA.UL.AAZ6.01 & TCGA.UU.A93S.01\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &  450.655142 & 3020.109054 & 6486.0687   & 3641.6943   &  875.8434   & 5252.389068 & 4349.910381 & 3964.5744   & 2835.217596 & 4355.0541   & ⋯           &  645.52538  & 4536.742724 & 3691.4786   & 2400.522368 & 1371.52306  & 1661.2877   & 2090.7229   & 1186.40144  &  976.8824   &  364.1602  \\\\\n",
       "\tENSG00000000005 &    2.000078 &    3.999903 &    0.0000   &    1.0000   &    1.0000   &    4.999912 &    6.000185 &    3.0000   &    3.999903 &  731.0001   & ⋯           &   11.00031  &    6.000219 & 3556.0146   &    5.000156 &   24.99974  &    0.0000   &    7.0000   &   37.99979  &    1.0000   &   36.0012  \\\\\n",
       "\tENSG00000000419 & 1821.031845 & 1405.018826 & 3620.0691   & 2324.9518   & 5508.9806   & 4677.870576 & 2524.012514 &  873.9736   & 1888.007985 & 3316.0325   & ⋯           & 1597.98718  & 2047.962746 & 1358.0296   & 1512.008500 & 1189.99584  & 2402.0106   & 1015.0289   & 1221.00364  & 1163.9743   & 1328.0154  \\\\\n",
       "\tENSG00000000457 & 3688.499306 & 1940.060272 & 1323.6190   & 1972.4032   & 2573.8937   & 1543.572332 & 1095.212755 & 1096.8912   & 2412.455125 & 3308.7035   & ⋯           &  710.14563  & 1259.116073 &  754.7742   & 1227.154984 & 2042.30419  &  549.9404   & 1087.0107   &  961.86779  & 1708.3689   & 1660.7872  \\\\\n",
       "\tENSG00000000460 &  769.629818 &  743.029409 & 1769.5213   & 1138.3500   & 1300.2310   & 2156.624673 & 1349.009842 &  820.7999   & 1170.767933 & 2410.5476   & ⋯           &  528.78813  & 1924.489990 &  364.9363   & 1044.840011 & 1096.15716  &  725.6080   &  509.8673   & 1310.12223  &  856.8481   &  957.7056  \\\\\n",
       "\tENSG00000000938 &  383.997866 &  385.992250 &  115.9989   &  689.0194   &  411.9968   &  191.001862 &  924.994900 &  381.0095   & 1708.974047 &  839.0132   & ⋯           &  685.98680  &  461.996608 &  185.0003   &  310.004006 &  478.99150  & 1183.9836   &  169.0000   &  955.01035  &  215.9941   &  186.0013  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A1.A0SO.01 | TCGA.A1.A0SP.01 | TCGA.A2.A04P.01 | TCGA.A2.A04Q.01 | TCGA.A2.A04T.01 | ⋯ | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.PE.A5DC.01 | TCGA.PE.A5DD.01 | TCGA.S3.AA10.01 | TCGA.S3.AA14.01 | TCGA.S3.AA15.01 | TCGA.UL.AAZ6.01 | TCGA.UU.A93S.01 | \n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |  450.655142 | 3020.109054 | 6486.0687   | 3641.6943   |  875.8434   | 5252.389068 | 4349.910381 | 3964.5744   | 2835.217596 | 4355.0541   | ⋯           |  645.52538  | 4536.742724 | 3691.4786   | 2400.522368 | 1371.52306  | 1661.2877   | 2090.7229   | 1186.40144  |  976.8824   |  364.1602   | \n",
       "| ENSG00000000005 |    2.000078 |    3.999903 |    0.0000   |    1.0000   |    1.0000   |    4.999912 |    6.000185 |    3.0000   |    3.999903 |  731.0001   | ⋯           |   11.00031  |    6.000219 | 3556.0146   |    5.000156 |   24.99974  |    0.0000   |    7.0000   |   37.99979  |    1.0000   |   36.0012   | \n",
       "| ENSG00000000419 | 1821.031845 | 1405.018826 | 3620.0691   | 2324.9518   | 5508.9806   | 4677.870576 | 2524.012514 |  873.9736   | 1888.007985 | 3316.0325   | ⋯           | 1597.98718  | 2047.962746 | 1358.0296   | 1512.008500 | 1189.99584  | 2402.0106   | 1015.0289   | 1221.00364  | 1163.9743   | 1328.0154   | \n",
       "| ENSG00000000457 | 3688.499306 | 1940.060272 | 1323.6190   | 1972.4032   | 2573.8937   | 1543.572332 | 1095.212755 | 1096.8912   | 2412.455125 | 3308.7035   | ⋯           |  710.14563  | 1259.116073 |  754.7742   | 1227.154984 | 2042.30419  |  549.9404   | 1087.0107   |  961.86779  | 1708.3689   | 1660.7872   | \n",
       "| ENSG00000000460 |  769.629818 |  743.029409 | 1769.5213   | 1138.3500   | 1300.2310   | 2156.624673 | 1349.009842 |  820.7999   | 1170.767933 | 2410.5476   | ⋯           |  528.78813  | 1924.489990 |  364.9363   | 1044.840011 | 1096.15716  |  725.6080   |  509.8673   | 1310.12223  |  856.8481   |  957.7056   | \n",
       "| ENSG00000000938 |  383.997866 |  385.992250 |  115.9989   |  689.0194   |  411.9968   |  191.001862 |  924.994900 |  381.0095   | 1708.974047 |  839.0132   | ⋯           |  685.98680  |  461.996608 |  185.0003   |  310.004006 |  478.99150  | 1183.9836   |  169.0000   |  955.01035  |  215.9941   |  186.0013   | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SK.01 TCGA.A1.A0SM.01\n",
       "ENSG00000000003  450.655142     3020.109054     6486.0687       3641.6943      \n",
       "ENSG00000000005    2.000078        3.999903        0.0000          1.0000      \n",
       "ENSG00000000419 1821.031845     1405.018826     3620.0691       2324.9518      \n",
       "ENSG00000000457 3688.499306     1940.060272     1323.6190       1972.4032      \n",
       "ENSG00000000460  769.629818      743.029409     1769.5213       1138.3500      \n",
       "ENSG00000000938  383.997866      385.992250      115.9989        689.0194      \n",
       "                TCGA.A1.A0SN.01 TCGA.A1.A0SO.01 TCGA.A1.A0SP.01 TCGA.A2.A04P.01\n",
       "ENSG00000000003  875.8434       5252.389068     4349.910381     3964.5744      \n",
       "ENSG00000000005    1.0000          4.999912        6.000185        3.0000      \n",
       "ENSG00000000419 5508.9806       4677.870576     2524.012514      873.9736      \n",
       "ENSG00000000457 2573.8937       1543.572332     1095.212755     1096.8912      \n",
       "ENSG00000000460 1300.2310       2156.624673     1349.009842      820.7999      \n",
       "ENSG00000000938  411.9968        191.001862      924.994900      381.0095      \n",
       "                TCGA.A2.A04Q.01 TCGA.A2.A04T.01 ⋯ TCGA.OL.A66P.01\n",
       "ENSG00000000003 2835.217596     4355.0541       ⋯  645.52538     \n",
       "ENSG00000000005    3.999903      731.0001       ⋯   11.00031     \n",
       "ENSG00000000419 1888.007985     3316.0325       ⋯ 1597.98718     \n",
       "ENSG00000000457 2412.455125     3308.7035       ⋯  710.14563     \n",
       "ENSG00000000460 1170.767933     2410.5476       ⋯  528.78813     \n",
       "ENSG00000000938 1708.974047      839.0132       ⋯  685.98680     \n",
       "                TCGA.OL.A6VO.01 TCGA.OL.A97C.01 TCGA.PE.A5DC.01 TCGA.PE.A5DD.01\n",
       "ENSG00000000003 4536.742724     3691.4786       2400.522368     1371.52306     \n",
       "ENSG00000000005    6.000219     3556.0146          5.000156       24.99974     \n",
       "ENSG00000000419 2047.962746     1358.0296       1512.008500     1189.99584     \n",
       "ENSG00000000457 1259.116073      754.7742       1227.154984     2042.30419     \n",
       "ENSG00000000460 1924.489990      364.9363       1044.840011     1096.15716     \n",
       "ENSG00000000938  461.996608      185.0003        310.004006      478.99150     \n",
       "                TCGA.S3.AA10.01 TCGA.S3.AA14.01 TCGA.S3.AA15.01 TCGA.UL.AAZ6.01\n",
       "ENSG00000000003 1661.2877       2090.7229       1186.40144       976.8824      \n",
       "ENSG00000000005    0.0000          7.0000         37.99979         1.0000      \n",
       "ENSG00000000419 2402.0106       1015.0289       1221.00364      1163.9743      \n",
       "ENSG00000000457  549.9404       1087.0107        961.86779      1708.3689      \n",
       "ENSG00000000460  725.6080        509.8673       1310.12223       856.8481      \n",
       "ENSG00000000938 1183.9836        169.0000        955.01035       215.9941      \n",
       "                TCGA.UU.A93S.01\n",
       "ENSG00000000003  364.1602      \n",
       "ENSG00000000005   36.0012      \n",
       "ENSG00000000419 1328.0154      \n",
       "ENSG00000000457 1660.7872      \n",
       "ENSG00000000460  957.7056      \n",
       "ENSG00000000938  186.0013      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#we need to make sure that TCGA ids are the same in the clinical and counts data\n",
    "head(brca.clin.selected)\n",
    "head(brca.untran.gene.counts)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** that the IDs are slightly different. In the clinical data the IDs contain \"-\" while the counts data IDs contain periods. We need to correct this in our dataframe explaining our samples. We can do that by simply subbing \".\" for \"-\" in the clinical data (to match the counts data) using gsub with the identifiers of interest. For us, the IDs of interest are HER2 and TNBC samples that also have counts data. We do this with the blocks of code below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "old clinical data names\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'TCGA-3C-AAAU-01'</li>\n",
       "\t<li>'TCGA-3C-AALI-01'</li>\n",
       "\t<li>'TCGA-3C-AALJ-01'</li>\n",
       "\t<li>'TCGA-3C-AALK-01'</li>\n",
       "\t<li>'TCGA-4H-AAAK-01'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TCGA-3C-AAAU-01'\n",
       "\\item 'TCGA-3C-AALI-01'\n",
       "\\item 'TCGA-3C-AALJ-01'\n",
       "\\item 'TCGA-3C-AALK-01'\n",
       "\\item 'TCGA-4H-AAAK-01'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TCGA-3C-AAAU-01'\n",
       "2. 'TCGA-3C-AALI-01'\n",
       "3. 'TCGA-3C-AALJ-01'\n",
       "4. 'TCGA-3C-AALK-01'\n",
       "5. 'TCGA-4H-AAAK-01'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"TCGA-3C-AAAU-01\" \"TCGA-3C-AALI-01\" \"TCGA-3C-AALJ-01\" \"TCGA-3C-AALK-01\"\n",
       "[5] \"TCGA-4H-AAAK-01\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converted  HER2 IDs\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'TCGA.3C.AALI.01'</li>\n",
       "\t<li>'TCGA.3C.AALK.01'</li>\n",
       "\t<li>'TCGA.A1.A0SM.01'</li>\n",
       "\t<li>'TCGA.A1.A0SN.01'</li>\n",
       "\t<li>'TCGA.A2.A04W.01'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TCGA.3C.AALI.01'\n",
       "\\item 'TCGA.3C.AALK.01'\n",
       "\\item 'TCGA.A1.A0SM.01'\n",
       "\\item 'TCGA.A1.A0SN.01'\n",
       "\\item 'TCGA.A2.A04W.01'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TCGA.3C.AALI.01'\n",
       "2. 'TCGA.3C.AALK.01'\n",
       "3. 'TCGA.A1.A0SM.01'\n",
       "4. 'TCGA.A1.A0SN.01'\n",
       "5. 'TCGA.A2.A04W.01'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"TCGA.3C.AALI.01\" \"TCGA.3C.AALK.01\" \"TCGA.A1.A0SM.01\" \"TCGA.A1.A0SN.01\"\n",
       "[5] \"TCGA.A2.A04W.01\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs from counts data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'TCGA.3C.AALI.01'</li>\n",
       "\t<li>'TCGA.3C.AALK.01'</li>\n",
       "\t<li>'TCGA.A1.A0SK.01'</li>\n",
       "\t<li>'TCGA.A1.A0SM.01'</li>\n",
       "\t<li>'TCGA.A1.A0SN.01'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TCGA.3C.AALI.01'\n",
       "\\item 'TCGA.3C.AALK.01'\n",
       "\\item 'TCGA.A1.A0SK.01'\n",
       "\\item 'TCGA.A1.A0SM.01'\n",
       "\\item 'TCGA.A1.A0SN.01'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TCGA.3C.AALI.01'\n",
       "2. 'TCGA.3C.AALK.01'\n",
       "3. 'TCGA.A1.A0SK.01'\n",
       "4. 'TCGA.A1.A0SM.01'\n",
       "5. 'TCGA.A1.A0SN.01'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"TCGA.3C.AALI.01\" \"TCGA.3C.AALK.01\" \"TCGA.A1.A0SK.01\" \"TCGA.A1.A0SM.01\"\n",
       "[5] \"TCGA.A1.A0SN.01\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#note that the IDs are slightly different. In the clinical data the IDs contain \"-\" while in the counts data IDs contain \".\"\n",
    "#we will need to rectify when running DESeq2\n",
    "\n",
    "#make vectors of all HER2 positive and TNBC IDs from the clinical data \n",
    "#and change - to . to match brca.untran.gene.counts (output of tximport) \n",
    "her2.ids <- as.vector(gsub(\"-\", \".\", brca.clin.selected[brca.clin.selected$HER2 == \"Positive\", 1]))\n",
    "tnbc.ids <- as.vector(gsub(\"-\", \".\", brca.clin.selected[brca.clin.selected$TNBC == \"TNBC\", 1]))\n",
    "\n",
    "#examine the results of the gsub to confirm it worked\n",
    "\n",
    "#old clinical data names\n",
    "message(\"old clinical data names\")\n",
    "brca.clin.selected[1:5 ,1]\n",
    "\n",
    "#converted her2 IDs\n",
    "message(\"converted  HER2 IDs\")\n",
    "her2.ids[1:5]\n",
    "\n",
    "#list of IDs from counts data\n",
    "message(\"IDs from counts data\")\n",
    "colnames(brca.untran.gene.counts)[1:5]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note** the IDs now match, but remember that not all HER2+ and TNBC samples have RNAseq data. If we intersect the column names from the counts data with the converted HER2+ and TNBC IDs, we will identify all identifiers with counts data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "number of HER2 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "206"
      ],
      "text/latex": [
       "206"
      ],
      "text/markdown": [
       "206"
      ],
      "text/plain": [
       "[1] 206"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "number of HER2 samples with counts data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "181"
      ],
      "text/latex": [
       "181"
      ],
      "text/markdown": [
       "181"
      ],
      "text/plain": [
       "[1] 181"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "number of TNBC samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "176"
      ],
      "text/latex": [
       "176"
      ],
      "text/markdown": [
       "176"
      ],
      "text/plain": [
       "[1] 176"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "number of TNBC samples with counts data\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "157"
      ],
      "text/latex": [
       "157"
      ],
      "text/markdown": [
       "157"
      ],
      "text/plain": [
       "[1] 157"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "number of HER2 and TNBC samples with counts\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "338"
      ],
      "text/latex": [
       "338"
      ],
      "text/markdown": [
       "338"
      ],
      "text/plain": [
       "[1] 338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#it is possible to have clinical data without rnaseq data so we need to find patient samples that have both\n",
    "her2.ids.final <- intersect(her2.ids, colnames(brca.untran.gene.counts))\n",
    "tnbc.ids.final <- intersect(tnbc.ids, colnames(brca.untran.gene.counts))\n",
    "\n",
    "message(\"number of HER2 samples\")\n",
    "length(her2.ids)\n",
    "\n",
    "message(\"number of HER2 samples with counts data\")\n",
    "length(her2.ids.final)\n",
    "\n",
    "message(\"number of TNBC samples\")\n",
    "length(tnbc.ids)\n",
    "\n",
    "message(\"number of TNBC samples with counts data\")\n",
    "length(tnbc.ids.final)\n",
    "\n",
    "message(\"number of HER2 and TNBC samples with counts\")\n",
    "length(her2.ids.final) + length(tnbc.ids.final)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we will need these IDs later, so save them to file\n",
    "save(her2.ids.final, file = file.path(date, \"her2.ids.final.Rda\"))\n",
    "save(tnbc.ids.final, file = file.path(date, \"tnbc.ids.final.Rda\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is positive reaffirmation that everything is working. As in Notebook 2, we have 206 HER2+ IDs and 176 TNBC IDs. Also, the number of HER2 and TNBC IDs with counts data sums to 338, as it did in Notebook 2. Now we can create what is necessary to run DESeq2:\n",
    "\n",
    "1. A dataframe explaining all samples (TCGA IDs) and their perscribed \"group.\" For ex. TCGA.XX.XXXY.01\tis TNBC and TCGA.XX.XXXX.01 is HER2.\n",
    "2. The counts data with TCGA identifiers that are exactly the same as the dataframe above. These values also must be integers to run DESeq2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Create the counts dataframe for DESeq2\n",
    "The counts dataframe is very straightforward to create now that the IDs match perfectly. We will subset the entire gene.counts data frame to create HER2 positive and TNBC-specific dataframes of the counts. We will then bind them together and floor because DESeq2 expects integers. You could also round the counts, it will not make a noticable difference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>TCGA.3C.AALI.01</th><th scope=col>TCGA.3C.AALK.01</th><th scope=col>TCGA.A1.A0SM.01</th><th scope=col>TCGA.A1.A0SN.01</th><th scope=col>TCGA.A2.A04W.01</th><th scope=col>TCGA.A2.A04X.01</th><th scope=col>TCGA.A2.A0CX.01</th><th scope=col>TCGA.A2.A0D1.01</th><th scope=col>TCGA.A2.A0EY.01</th><th scope=col>TCGA.A2.A0SY.01</th><th scope=col>⋯</th><th scope=col>TCGA.LL.A740.01</th><th scope=col>TCGA.OL.A5D6.01</th><th scope=col>TCGA.OL.A5D7.01</th><th scope=col>TCGA.OL.A5RW.01</th><th scope=col>TCGA.OL.A66I.01</th><th scope=col>TCGA.OL.A66P.01</th><th scope=col>TCGA.OL.A6VO.01</th><th scope=col>TCGA.OL.A97C.01</th><th scope=col>TCGA.S3.AA10.01</th><th scope=col>TCGA.S3.AA15.01</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000000003</th><td> 450</td><td>3020</td><td>3641</td><td> 875</td><td>2779</td><td>1429</td><td>3276</td><td>3980</td><td>2004</td><td>2056</td><td>⋯   </td><td>1489</td><td>1434</td><td>2034</td><td>1249</td><td>2389</td><td> 645</td><td>4536</td><td>3691</td><td>1661</td><td>1186</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000005</th><td>   2</td><td>   3</td><td>   1</td><td>   1</td><td>   1</td><td>   3</td><td>   1</td><td>   8</td><td>  15</td><td>  70</td><td>⋯   </td><td>   1</td><td>   2</td><td>   4</td><td>   0</td><td>  56</td><td>  11</td><td>   6</td><td>3556</td><td>   0</td><td>  37</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000419</th><td>1821</td><td>1405</td><td>2324</td><td>5508</td><td>1935</td><td>3256</td><td>1924</td><td>1219</td><td>2143</td><td>1810</td><td>⋯   </td><td>2093</td><td>1141</td><td>1880</td><td>1544</td><td>3579</td><td>1597</td><td>2047</td><td>1358</td><td>2402</td><td>1221</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000457</th><td>3688</td><td>1940</td><td>1972</td><td>2573</td><td>1088</td><td>1336</td><td>2156</td><td>1231</td><td>2562</td><td>2314</td><td>⋯   </td><td>1130</td><td>1342</td><td>1737</td><td>1282</td><td>1986</td><td> 710</td><td>1259</td><td> 754</td><td> 549</td><td> 961</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000460</th><td> 769</td><td> 743</td><td>1138</td><td>1300</td><td> 503</td><td> 508</td><td>1157</td><td> 751</td><td>1020</td><td> 851</td><td>⋯   </td><td> 654</td><td> 387</td><td>2625</td><td>1859</td><td>2529</td><td> 528</td><td>1924</td><td> 364</td><td> 725</td><td>1310</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000000938</th><td> 383</td><td> 385</td><td> 689</td><td> 411</td><td> 226</td><td> 869</td><td> 344</td><td> 180</td><td> 310</td><td> 510</td><td>⋯   </td><td> 730</td><td> 534</td><td> 689</td><td> 189</td><td>1277</td><td> 685</td><td> 461</td><td> 185</td><td>1183</td><td> 955</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllll}\n",
       "  & TCGA.3C.AALI.01 & TCGA.3C.AALK.01 & TCGA.A1.A0SM.01 & TCGA.A1.A0SN.01 & TCGA.A2.A04W.01 & TCGA.A2.A04X.01 & TCGA.A2.A0CX.01 & TCGA.A2.A0D1.01 & TCGA.A2.A0EY.01 & TCGA.A2.A0SY.01 & ⋯ & TCGA.LL.A740.01 & TCGA.OL.A5D6.01 & TCGA.OL.A5D7.01 & TCGA.OL.A5RW.01 & TCGA.OL.A66I.01 & TCGA.OL.A66P.01 & TCGA.OL.A6VO.01 & TCGA.OL.A97C.01 & TCGA.S3.AA10.01 & TCGA.S3.AA15.01\\\\\n",
       "\\hline\n",
       "\tENSG00000000003 &  450 & 3020 & 3641 &  875 & 2779 & 1429 & 3276 & 3980 & 2004 & 2056 & ⋯    & 1489 & 1434 & 2034 & 1249 & 2389 &  645 & 4536 & 3691 & 1661 & 1186\\\\\n",
       "\tENSG00000000005 &    2 &    3 &    1 &    1 &    1 &    3 &    1 &    8 &   15 &   70 & ⋯    &    1 &    2 &    4 &    0 &   56 &   11 &    6 & 3556 &    0 &   37\\\\\n",
       "\tENSG00000000419 & 1821 & 1405 & 2324 & 5508 & 1935 & 3256 & 1924 & 1219 & 2143 & 1810 & ⋯    & 2093 & 1141 & 1880 & 1544 & 3579 & 1597 & 2047 & 1358 & 2402 & 1221\\\\\n",
       "\tENSG00000000457 & 3688 & 1940 & 1972 & 2573 & 1088 & 1336 & 2156 & 1231 & 2562 & 2314 & ⋯    & 1130 & 1342 & 1737 & 1282 & 1986 &  710 & 1259 &  754 &  549 &  961\\\\\n",
       "\tENSG00000000460 &  769 &  743 & 1138 & 1300 &  503 &  508 & 1157 &  751 & 1020 &  851 & ⋯    &  654 &  387 & 2625 & 1859 & 2529 &  528 & 1924 &  364 &  725 & 1310\\\\\n",
       "\tENSG00000000938 &  383 &  385 &  689 &  411 &  226 &  869 &  344 &  180 &  310 &  510 & ⋯    &  730 &  534 &  689 &  189 & 1277 &  685 &  461 &  185 & 1183 &  955\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | TCGA.3C.AALI.01 | TCGA.3C.AALK.01 | TCGA.A1.A0SM.01 | TCGA.A1.A0SN.01 | TCGA.A2.A04W.01 | TCGA.A2.A04X.01 | TCGA.A2.A0CX.01 | TCGA.A2.A0D1.01 | TCGA.A2.A0EY.01 | TCGA.A2.A0SY.01 | ⋯ | TCGA.LL.A740.01 | TCGA.OL.A5D6.01 | TCGA.OL.A5D7.01 | TCGA.OL.A5RW.01 | TCGA.OL.A66I.01 | TCGA.OL.A66P.01 | TCGA.OL.A6VO.01 | TCGA.OL.A97C.01 | TCGA.S3.AA10.01 | TCGA.S3.AA15.01 | \n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000000003 |  450 | 3020 | 3641 |  875 | 2779 | 1429 | 3276 | 3980 | 2004 | 2056 | ⋯    | 1489 | 1434 | 2034 | 1249 | 2389 |  645 | 4536 | 3691 | 1661 | 1186 | \n",
       "| ENSG00000000005 |    2 |    3 |    1 |    1 |    1 |    3 |    1 |    8 |   15 |   70 | ⋯    |    1 |    2 |    4 |    0 |   56 |   11 |    6 | 3556 |    0 |   37 | \n",
       "| ENSG00000000419 | 1821 | 1405 | 2324 | 5508 | 1935 | 3256 | 1924 | 1219 | 2143 | 1810 | ⋯    | 2093 | 1141 | 1880 | 1544 | 3579 | 1597 | 2047 | 1358 | 2402 | 1221 | \n",
       "| ENSG00000000457 | 3688 | 1940 | 1972 | 2573 | 1088 | 1336 | 2156 | 1231 | 2562 | 2314 | ⋯    | 1130 | 1342 | 1737 | 1282 | 1986 |  710 | 1259 |  754 |  549 |  961 | \n",
       "| ENSG00000000460 |  769 |  743 | 1138 | 1300 |  503 |  508 | 1157 |  751 | 1020 |  851 | ⋯    |  654 |  387 | 2625 | 1859 | 2529 |  528 | 1924 |  364 |  725 | 1310 | \n",
       "| ENSG00000000938 |  383 |  385 |  689 |  411 |  226 |  869 |  344 |  180 |  310 |  510 | ⋯    |  730 |  534 |  689 |  189 | 1277 |  685 |  461 |  185 | 1183 |  955 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                TCGA.3C.AALI.01 TCGA.3C.AALK.01 TCGA.A1.A0SM.01 TCGA.A1.A0SN.01\n",
       "ENSG00000000003  450            3020            3641             875           \n",
       "ENSG00000000005    2               3               1               1           \n",
       "ENSG00000000419 1821            1405            2324            5508           \n",
       "ENSG00000000457 3688            1940            1972            2573           \n",
       "ENSG00000000460  769             743            1138            1300           \n",
       "ENSG00000000938  383             385             689             411           \n",
       "                TCGA.A2.A04W.01 TCGA.A2.A04X.01 TCGA.A2.A0CX.01 TCGA.A2.A0D1.01\n",
       "ENSG00000000003 2779            1429            3276            3980           \n",
       "ENSG00000000005    1               3               1               8           \n",
       "ENSG00000000419 1935            3256            1924            1219           \n",
       "ENSG00000000457 1088            1336            2156            1231           \n",
       "ENSG00000000460  503             508            1157             751           \n",
       "ENSG00000000938  226             869             344             180           \n",
       "                TCGA.A2.A0EY.01 TCGA.A2.A0SY.01 ⋯ TCGA.LL.A740.01\n",
       "ENSG00000000003 2004            2056            ⋯ 1489           \n",
       "ENSG00000000005   15              70            ⋯    1           \n",
       "ENSG00000000419 2143            1810            ⋯ 2093           \n",
       "ENSG00000000457 2562            2314            ⋯ 1130           \n",
       "ENSG00000000460 1020             851            ⋯  654           \n",
       "ENSG00000000938  310             510            ⋯  730           \n",
       "                TCGA.OL.A5D6.01 TCGA.OL.A5D7.01 TCGA.OL.A5RW.01 TCGA.OL.A66I.01\n",
       "ENSG00000000003 1434            2034            1249            2389           \n",
       "ENSG00000000005    2               4               0              56           \n",
       "ENSG00000000419 1141            1880            1544            3579           \n",
       "ENSG00000000457 1342            1737            1282            1986           \n",
       "ENSG00000000460  387            2625            1859            2529           \n",
       "ENSG00000000938  534             689             189            1277           \n",
       "                TCGA.OL.A66P.01 TCGA.OL.A6VO.01 TCGA.OL.A97C.01 TCGA.S3.AA10.01\n",
       "ENSG00000000003  645            4536            3691            1661           \n",
       "ENSG00000000005   11               6            3556               0           \n",
       "ENSG00000000419 1597            2047            1358            2402           \n",
       "ENSG00000000457  710            1259             754             549           \n",
       "ENSG00000000460  528            1924             364             725           \n",
       "ENSG00000000938  685             461             185            1183           \n",
       "                TCGA.S3.AA15.01\n",
       "ENSG00000000003 1186           \n",
       "ENSG00000000005   37           \n",
       "ENSG00000000419 1221           \n",
       "ENSG00000000457  961           \n",
       "ENSG00000000460 1310           \n",
       "ENSG00000000938  955           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "dimensions of counts data frame\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>49922</li>\n",
       "\t<li>338</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 49922\n",
       "\\item 338\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 49922\n",
       "2. 338\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 49922   338"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "######################################################################################\n",
    "############### create counts dataframe for DESeq2 ###################################\n",
    "######################################################################################\n",
    "\n",
    "#now create counts dataframes of her2 and tnbc, bind together, and floor because they must be integers for DESeq2\n",
    "her2.counts.df <- brca.untran.gene.counts[, her2.ids.final] \n",
    "tnbc.counts.df <- brca.untran.gene.counts[, tnbc.ids.final]\n",
    "\n",
    "#bind these together and floor to make all values integers.\n",
    "her2.tnbc.counts.df <- floor(cbind(her2.counts.df, tnbc.counts.df))\n",
    "\n",
    "#examine the dataframe\n",
    "head(her2.tnbc.counts.df)\n",
    "\n",
    "#check the dimensions to check the correct number of samples and the number of genes\n",
    "message(\"dimensions of counts data frame\")\n",
    "dim(her2.tnbc.counts.df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** if the dataframe is correct it should contain 338 samples (as calculated above HER2 positive + TNBC samples with counts data) and 49922 genes (determined by tximport output)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Create knowledge dataframe for DESeq2\n",
    "The knowledge dataframe is simply a two column dataframe. Column one is the TCGA ID of each sample and column two is their corresponding breast cancer subtype; \"HER2\" or \"TNBC.\" There are multiple ways to create the dataframe. We will take the simple approach and create a vector that will specify breast cancer subtype by repeating \"HER2\" by the number of HER2 samples and \"TNBC\" by the number of TNBC samples. We calculated those numbers in 2.1 above. We will then append that vector to a vector of sample names from the final counts dataframe. **note** it is critical that we join the HER2 and TNBC vectors the same way we joined the counts dataframes above (HER2 then TNBC). We should then check that this is correct. Finally, we also will specify each row name as the same sample name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>sample</th><th scope=col>BC_subtype</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>TCGA.3C.AALI.01</th><td>TCGA.3C.AALI.01</td><td>HER2           </td></tr>\n",
       "\t<tr><th scope=row>TCGA.3C.AALK.01</th><td>TCGA.3C.AALK.01</td><td>HER2           </td></tr>\n",
       "\t<tr><th scope=row>TCGA.A1.A0SM.01</th><td>TCGA.A1.A0SM.01</td><td>HER2           </td></tr>\n",
       "\t<tr><th scope=row>TCGA.A1.A0SN.01</th><td>TCGA.A1.A0SN.01</td><td>HER2           </td></tr>\n",
       "\t<tr><th scope=row>TCGA.A2.A04W.01</th><td>TCGA.A2.A04W.01</td><td>HER2           </td></tr>\n",
       "\t<tr><th scope=row>TCGA.A2.A04X.01</th><td>TCGA.A2.A04X.01</td><td>HER2           </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       "  & sample & BC\\_subtype\\\\\n",
       "\\hline\n",
       "\tTCGA.3C.AALI.01 & TCGA.3C.AALI.01 & HER2           \\\\\n",
       "\tTCGA.3C.AALK.01 & TCGA.3C.AALK.01 & HER2           \\\\\n",
       "\tTCGA.A1.A0SM.01 & TCGA.A1.A0SM.01 & HER2           \\\\\n",
       "\tTCGA.A1.A0SN.01 & TCGA.A1.A0SN.01 & HER2           \\\\\n",
       "\tTCGA.A2.A04W.01 & TCGA.A2.A04W.01 & HER2           \\\\\n",
       "\tTCGA.A2.A04X.01 & TCGA.A2.A04X.01 & HER2           \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | sample | BC_subtype | \n",
       "|---|---|---|---|---|---|\n",
       "| TCGA.3C.AALI.01 | TCGA.3C.AALI.01 | HER2            | \n",
       "| TCGA.3C.AALK.01 | TCGA.3C.AALK.01 | HER2            | \n",
       "| TCGA.A1.A0SM.01 | TCGA.A1.A0SM.01 | HER2            | \n",
       "| TCGA.A1.A0SN.01 | TCGA.A1.A0SN.01 | HER2            | \n",
       "| TCGA.A2.A04W.01 | TCGA.A2.A04W.01 | HER2            | \n",
       "| TCGA.A2.A04X.01 | TCGA.A2.A04X.01 | HER2            | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                sample          BC_subtype\n",
       "TCGA.3C.AALI.01 TCGA.3C.AALI.01 HER2      \n",
       "TCGA.3C.AALK.01 TCGA.3C.AALK.01 HER2      \n",
       "TCGA.A1.A0SM.01 TCGA.A1.A0SM.01 HER2      \n",
       "TCGA.A1.A0SN.01 TCGA.A1.A0SN.01 HER2      \n",
       "TCGA.A2.A04W.01 TCGA.A2.A04W.01 HER2      \n",
       "TCGA.A2.A04X.01 TCGA.A2.A04X.01 HER2      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#########################################################################################\n",
    "################### create knowledge file for DESeq2 see manual ##########################\n",
    "##########################################################################################\n",
    "\n",
    "numHER2 <- ncol(her2.counts.df)\n",
    "numTNBC <- ncol(tnbc.counts.df)\n",
    "\n",
    "#BC_subtype defines what columns are HER2 samples vs TNBC samples\n",
    "BC_subtype <- c(rep(\"HER2\", numHER2), rep(\"TNBC\", numTNBC))\n",
    "\n",
    "#samples of interest are merely the name of the samples that we will examine\n",
    "samplesofinterest <- colnames(her2.tnbc.counts.df)\n",
    "\n",
    "#sampleTable is the combination of the above two\n",
    "#it is a dataframe with sample name and if it is HER2 or TNBC\n",
    "sampleTable = data.frame(sample = samplesofinterest, BC_subtype = BC_subtype, row.names = samplesofinterest)\n",
    "\n",
    "#examine the sampleTable\n",
    "head(sampleTable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the dataframe is in the correct format, you should see a dataframe with rownames that are the sample name and two columns -'sample' and'BC_subtype'. Let's confirm the samples are correctly labeled by making sure that the samples labled HER2 and TNBC in the sampleTable correspond to the HER2 and TNBC IDs we calculated above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "HER2 logic\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. TRUE\n",
       "2. TRUE\n",
       "3. TRUE\n",
       "4. TRUE\n",
       "5. TRUE\n",
       "6. TRUE\n",
       "7. TRUE\n",
       "8. TRUE\n",
       "9. TRUE\n",
       "10. TRUE\n",
       "11. TRUE\n",
       "12. TRUE\n",
       "13. TRUE\n",
       "14. TRUE\n",
       "15. TRUE\n",
       "16. TRUE\n",
       "17. TRUE\n",
       "18. TRUE\n",
       "19. TRUE\n",
       "20. TRUE\n",
       "21. TRUE\n",
       "22. TRUE\n",
       "23. TRUE\n",
       "24. TRUE\n",
       "25. TRUE\n",
       "26. TRUE\n",
       "27. TRUE\n",
       "28. TRUE\n",
       "29. TRUE\n",
       "30. TRUE\n",
       "31. TRUE\n",
       "32. TRUE\n",
       "33. TRUE\n",
       "34. TRUE\n",
       "35. TRUE\n",
       "36. TRUE\n",
       "37. TRUE\n",
       "38. TRUE\n",
       "39. TRUE\n",
       "40. TRUE\n",
       "41. TRUE\n",
       "42. TRUE\n",
       "43. TRUE\n",
       "44. TRUE\n",
       "45. TRUE\n",
       "46. TRUE\n",
       "47. TRUE\n",
       "48. TRUE\n",
       "49. TRUE\n",
       "50. TRUE\n",
       "51. TRUE\n",
       "52. TRUE\n",
       "53. TRUE\n",
       "54. TRUE\n",
       "55. TRUE\n",
       "56. TRUE\n",
       "57. TRUE\n",
       "58. TRUE\n",
       "59. TRUE\n",
       "60. TRUE\n",
       "61. TRUE\n",
       "62. TRUE\n",
       "63. TRUE\n",
       "64. TRUE\n",
       "65. TRUE\n",
       "66. TRUE\n",
       "67. TRUE\n",
       "68. TRUE\n",
       "69. TRUE\n",
       "70. TRUE\n",
       "71. TRUE\n",
       "72. TRUE\n",
       "73. TRUE\n",
       "74. TRUE\n",
       "75. TRUE\n",
       "76. TRUE\n",
       "77. TRUE\n",
       "78. TRUE\n",
       "79. TRUE\n",
       "80. TRUE\n",
       "81. TRUE\n",
       "82. TRUE\n",
       "83. TRUE\n",
       "84. TRUE\n",
       "85. TRUE\n",
       "86. TRUE\n",
       "87. TRUE\n",
       "88. TRUE\n",
       "89. TRUE\n",
       "90. TRUE\n",
       "91. TRUE\n",
       "92. TRUE\n",
       "93. TRUE\n",
       "94. TRUE\n",
       "95. TRUE\n",
       "96. TRUE\n",
       "97. TRUE\n",
       "98. TRUE\n",
       "99. TRUE\n",
       "100. TRUE\n",
       "101. TRUE\n",
       "102. TRUE\n",
       "103. TRUE\n",
       "104. TRUE\n",
       "105. TRUE\n",
       "106. TRUE\n",
       "107. TRUE\n",
       "108. TRUE\n",
       "109. TRUE\n",
       "110. TRUE\n",
       "111. TRUE\n",
       "112. TRUE\n",
       "113. TRUE\n",
       "114. TRUE\n",
       "115. TRUE\n",
       "116. TRUE\n",
       "117. TRUE\n",
       "118. TRUE\n",
       "119. TRUE\n",
       "120. TRUE\n",
       "121. TRUE\n",
       "122. TRUE\n",
       "123. TRUE\n",
       "124. TRUE\n",
       "125. TRUE\n",
       "126. TRUE\n",
       "127. TRUE\n",
       "128. TRUE\n",
       "129. TRUE\n",
       "130. TRUE\n",
       "131. TRUE\n",
       "132. TRUE\n",
       "133. TRUE\n",
       "134. TRUE\n",
       "135. TRUE\n",
       "136. TRUE\n",
       "137. TRUE\n",
       "138. TRUE\n",
       "139. TRUE\n",
       "140. TRUE\n",
       "141. TRUE\n",
       "142. TRUE\n",
       "143. TRUE\n",
       "144. TRUE\n",
       "145. TRUE\n",
       "146. TRUE\n",
       "147. TRUE\n",
       "148. TRUE\n",
       "149. TRUE\n",
       "150. TRUE\n",
       "151. TRUE\n",
       "152. TRUE\n",
       "153. TRUE\n",
       "154. TRUE\n",
       "155. TRUE\n",
       "156. TRUE\n",
       "157. TRUE\n",
       "158. TRUE\n",
       "159. TRUE\n",
       "160. TRUE\n",
       "161. TRUE\n",
       "162. TRUE\n",
       "163. TRUE\n",
       "164. TRUE\n",
       "165. TRUE\n",
       "166. TRUE\n",
       "167. TRUE\n",
       "168. TRUE\n",
       "169. TRUE\n",
       "170. TRUE\n",
       "171. TRUE\n",
       "172. TRUE\n",
       "173. TRUE\n",
       "174. TRUE\n",
       "175. TRUE\n",
       "176. TRUE\n",
       "177. TRUE\n",
       "178. TRUE\n",
       "179. TRUE\n",
       "180. TRUE\n",
       "181. TRUE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [46] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [61] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [76] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [91] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[106] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[121] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[136] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[151] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[166] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[181] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TNBC logic\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "\t<li>TRUE</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\item TRUE\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. TRUE\n",
       "2. TRUE\n",
       "3. TRUE\n",
       "4. TRUE\n",
       "5. TRUE\n",
       "6. TRUE\n",
       "7. TRUE\n",
       "8. TRUE\n",
       "9. TRUE\n",
       "10. TRUE\n",
       "11. TRUE\n",
       "12. TRUE\n",
       "13. TRUE\n",
       "14. TRUE\n",
       "15. TRUE\n",
       "16. TRUE\n",
       "17. TRUE\n",
       "18. TRUE\n",
       "19. TRUE\n",
       "20. TRUE\n",
       "21. TRUE\n",
       "22. TRUE\n",
       "23. TRUE\n",
       "24. TRUE\n",
       "25. TRUE\n",
       "26. TRUE\n",
       "27. TRUE\n",
       "28. TRUE\n",
       "29. TRUE\n",
       "30. TRUE\n",
       "31. TRUE\n",
       "32. TRUE\n",
       "33. TRUE\n",
       "34. TRUE\n",
       "35. TRUE\n",
       "36. TRUE\n",
       "37. TRUE\n",
       "38. TRUE\n",
       "39. TRUE\n",
       "40. TRUE\n",
       "41. TRUE\n",
       "42. TRUE\n",
       "43. TRUE\n",
       "44. TRUE\n",
       "45. TRUE\n",
       "46. TRUE\n",
       "47. TRUE\n",
       "48. TRUE\n",
       "49. TRUE\n",
       "50. TRUE\n",
       "51. TRUE\n",
       "52. TRUE\n",
       "53. TRUE\n",
       "54. TRUE\n",
       "55. TRUE\n",
       "56. TRUE\n",
       "57. TRUE\n",
       "58. TRUE\n",
       "59. TRUE\n",
       "60. TRUE\n",
       "61. TRUE\n",
       "62. TRUE\n",
       "63. TRUE\n",
       "64. TRUE\n",
       "65. TRUE\n",
       "66. TRUE\n",
       "67. TRUE\n",
       "68. TRUE\n",
       "69. TRUE\n",
       "70. TRUE\n",
       "71. TRUE\n",
       "72. TRUE\n",
       "73. TRUE\n",
       "74. TRUE\n",
       "75. TRUE\n",
       "76. TRUE\n",
       "77. TRUE\n",
       "78. TRUE\n",
       "79. TRUE\n",
       "80. TRUE\n",
       "81. TRUE\n",
       "82. TRUE\n",
       "83. TRUE\n",
       "84. TRUE\n",
       "85. TRUE\n",
       "86. TRUE\n",
       "87. TRUE\n",
       "88. TRUE\n",
       "89. TRUE\n",
       "90. TRUE\n",
       "91. TRUE\n",
       "92. TRUE\n",
       "93. TRUE\n",
       "94. TRUE\n",
       "95. TRUE\n",
       "96. TRUE\n",
       "97. TRUE\n",
       "98. TRUE\n",
       "99. TRUE\n",
       "100. TRUE\n",
       "101. TRUE\n",
       "102. TRUE\n",
       "103. TRUE\n",
       "104. TRUE\n",
       "105. TRUE\n",
       "106. TRUE\n",
       "107. TRUE\n",
       "108. TRUE\n",
       "109. TRUE\n",
       "110. TRUE\n",
       "111. TRUE\n",
       "112. TRUE\n",
       "113. TRUE\n",
       "114. TRUE\n",
       "115. TRUE\n",
       "116. TRUE\n",
       "117. TRUE\n",
       "118. TRUE\n",
       "119. TRUE\n",
       "120. TRUE\n",
       "121. TRUE\n",
       "122. TRUE\n",
       "123. TRUE\n",
       "124. TRUE\n",
       "125. TRUE\n",
       "126. TRUE\n",
       "127. TRUE\n",
       "128. TRUE\n",
       "129. TRUE\n",
       "130. TRUE\n",
       "131. TRUE\n",
       "132. TRUE\n",
       "133. TRUE\n",
       "134. TRUE\n",
       "135. TRUE\n",
       "136. TRUE\n",
       "137. TRUE\n",
       "138. TRUE\n",
       "139. TRUE\n",
       "140. TRUE\n",
       "141. TRUE\n",
       "142. TRUE\n",
       "143. TRUE\n",
       "144. TRUE\n",
       "145. TRUE\n",
       "146. TRUE\n",
       "147. TRUE\n",
       "148. TRUE\n",
       "149. TRUE\n",
       "150. TRUE\n",
       "151. TRUE\n",
       "152. TRUE\n",
       "153. TRUE\n",
       "154. TRUE\n",
       "155. TRUE\n",
       "156. TRUE\n",
       "157. TRUE\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [46] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [61] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [76] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       " [91] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[106] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[121] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[136] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE\n",
       "[151] TRUE TRUE TRUE TRUE TRUE TRUE TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#subset dataframe to HER2 positive samples and check they are in vector of HER2 IDs\n",
    "#that we created from the clinical data\n",
    "message(\"HER2 logic\")\n",
    "sampleTable[sampleTable$BC_subtype == \"HER2\", 1] %in% her2.ids.final\n",
    "\n",
    "#do the same with TNBC IDs\n",
    "message(\"TNBC logic\")\n",
    "sampleTable[sampleTable$BC_subtype == \"TNBC\", 1] %in% tnbc.ids.final\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** If all samples are correctly labelled, the output from the preceeding code should all be TRUE "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4 Run DESeq2\n",
    "There are two steps to run DESeq2. The first (2.4a) is to create the object that DESeq2 uses for the statistical tests. The object class used by the DESeq2 package to store the read counts and the intermediate estimated quantities during statistical analysis is the DESeqDataSet. One caveat is that the counts data must be in a matrix instead of a dataframe. The second step (2.4b) is performing the statistical analysis, which is the most time consuming step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4a Creating the DESeqDataSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n"
     ]
    }
   ],
   "source": [
    "#########################################################################################\n",
    "################### Run the first step of DESeq2 ##########################\n",
    "##########################################################################################\n",
    "\n",
    "\n",
    "#her2.tnbc.counts.df is our final count matrix from her2 and tnbc samples only\n",
    "#sample table defines the subtpe for each sample\n",
    "#design in this case is simply HER2 vs TNBC, which is definied in BC_subtype of the sampleTable. This can be much more complicated depending on your samples\n",
    "#simply put, the design sets up the reference group for the differential expression analysis\n",
    "deseqdata <- DESeqDataSetFromMatrix(countData = as.matrix(her2.tnbc.counts.df), colData = sampleTable, design = ~ BC_subtype)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** Taken directly from the reference [manual](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseqdataset), \"By default, R will choose a reference level for factors based on alphabetical order. Meaning, if you do not specify which level you want to compare against (e.g. which level represents the control group), the comparisons will be based on the alphabetical order of the levels. We can manually set the comparison group using the relevel function on the deseqdata. You should only change the factor levels of variables in the design before running the DESeq2 analysis, not during or afterward.\" \n",
    "\n",
    "Accordingly, we will set the HER2 group as the reference label explicitly with the code below to avoid confusion when interpreting the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the relevel function sets HER2 as the reference level for fold change calculations\n",
    "deseqdata$BC_subtype <- relevel(deseqdata$BC_subtype, ref = \"HER2\")\n",
    "\n",
    "#now save deseqdata in case we ever need to start from here again\n",
    "save(deseqdata, file = file.path(date,\"her2vstnbc.deseq.Rda\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4b Perform the differential expression analysis with DESeq2\n",
    "This is the most time and computational consuming portion of this analysis. Before we begin, we will clear the global environment and RAM except for the necessary items."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "in global environment before rm\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'BC_subtype'</li>\n",
       "\t<li>'brca.clin.selected'</li>\n",
       "\t<li>'brca.untran.counts.df'</li>\n",
       "\t<li>'brca.untran.gene.counts'</li>\n",
       "\t<li>'brca.untran.gene.tpm'</li>\n",
       "\t<li>'brca.untran.tpm.df'</li>\n",
       "\t<li>'date'</li>\n",
       "\t<li>'deseqdata'</li>\n",
       "\t<li>'df.txi'</li>\n",
       "\t<li>'gene1_transcripts'</li>\n",
       "\t<li>'her2.counts.df'</li>\n",
       "\t<li>'her2.ids'</li>\n",
       "\t<li>'her2.ids.final'</li>\n",
       "\t<li>'her2.tnbc.counts.df'</li>\n",
       "\t<li>'numHER2'</li>\n",
       "\t<li>'numTNBC'</li>\n",
       "\t<li>'samplesofinterest'</li>\n",
       "\t<li>'sampleTable'</li>\n",
       "\t<li>'tnbc.counts.df'</li>\n",
       "\t<li>'tnbc.ids'</li>\n",
       "\t<li>'tnbc.ids.final'</li>\n",
       "\t<li>'tx2gene81'</li>\n",
       "\t<li>'txi'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BC\\_subtype'\n",
       "\\item 'brca.clin.selected'\n",
       "\\item 'brca.untran.counts.df'\n",
       "\\item 'brca.untran.gene.counts'\n",
       "\\item 'brca.untran.gene.tpm'\n",
       "\\item 'brca.untran.tpm.df'\n",
       "\\item 'date'\n",
       "\\item 'deseqdata'\n",
       "\\item 'df.txi'\n",
       "\\item 'gene1\\_transcripts'\n",
       "\\item 'her2.counts.df'\n",
       "\\item 'her2.ids'\n",
       "\\item 'her2.ids.final'\n",
       "\\item 'her2.tnbc.counts.df'\n",
       "\\item 'numHER2'\n",
       "\\item 'numTNBC'\n",
       "\\item 'samplesofinterest'\n",
       "\\item 'sampleTable'\n",
       "\\item 'tnbc.counts.df'\n",
       "\\item 'tnbc.ids'\n",
       "\\item 'tnbc.ids.final'\n",
       "\\item 'tx2gene81'\n",
       "\\item 'txi'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BC_subtype'\n",
       "2. 'brca.clin.selected'\n",
       "3. 'brca.untran.counts.df'\n",
       "4. 'brca.untran.gene.counts'\n",
       "5. 'brca.untran.gene.tpm'\n",
       "6. 'brca.untran.tpm.df'\n",
       "7. 'date'\n",
       "8. 'deseqdata'\n",
       "9. 'df.txi'\n",
       "10. 'gene1_transcripts'\n",
       "11. 'her2.counts.df'\n",
       "12. 'her2.ids'\n",
       "13. 'her2.ids.final'\n",
       "14. 'her2.tnbc.counts.df'\n",
       "15. 'numHER2'\n",
       "16. 'numTNBC'\n",
       "17. 'samplesofinterest'\n",
       "18. 'sampleTable'\n",
       "19. 'tnbc.counts.df'\n",
       "20. 'tnbc.ids'\n",
       "21. 'tnbc.ids.final'\n",
       "22. 'tx2gene81'\n",
       "23. 'txi'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"BC_subtype\"              \"brca.clin.selected\"     \n",
       " [3] \"brca.untran.counts.df\"   \"brca.untran.gene.counts\"\n",
       " [5] \"brca.untran.gene.tpm\"    \"brca.untran.tpm.df\"     \n",
       " [7] \"date\"                    \"deseqdata\"              \n",
       " [9] \"df.txi\"                  \"gene1_transcripts\"      \n",
       "[11] \"her2.counts.df\"          \"her2.ids\"               \n",
       "[13] \"her2.ids.final\"          \"her2.tnbc.counts.df\"    \n",
       "[15] \"numHER2\"                 \"numTNBC\"                \n",
       "[17] \"samplesofinterest\"       \"sampleTable\"            \n",
       "[19] \"tnbc.counts.df\"          \"tnbc.ids\"               \n",
       "[21] \"tnbc.ids.final\"          \"tx2gene81\"              \n",
       "[23] \"txi\"                    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>limit (Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Ncells</th><td> 5318280 </td><td>284.1    </td><td> 10272058</td><td> 548.6   </td><td>   NA    </td><td> 10272058</td><td> 548.6   </td></tr>\n",
       "\t<tr><th scope=row>Vcells</th><td>51950130 </td><td>396.4    </td><td>524288982</td><td>4000.1   </td><td>16384    </td><td>650032873</td><td>4959.4   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllll}\n",
       "  & used & (Mb) & gc trigger & (Mb) & limit (Mb) & max used & (Mb)\\\\\n",
       "\\hline\n",
       "\tNcells &  5318280  & 284.1     &  10272058 &  548.6    &    NA     &  10272058 &  548.6   \\\\\n",
       "\tVcells & 51950130  & 396.4     & 524288982 & 4000.1    & 16384     & 650032873 & 4959.4   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | used | (Mb) | gc trigger | (Mb) | limit (Mb) | max used | (Mb) | \n",
       "|---|---|\n",
       "| Ncells |  5318280  | 284.1     |  10272058 |  548.6    |    NA     |  10272058 |  548.6    | \n",
       "| Vcells | 51950130  | 396.4     | 524288982 | 4000.1    | 16384     | 650032873 | 4959.4    | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "       used     (Mb)  gc trigger (Mb)   limit (Mb) max used  (Mb)  \n",
       "Ncells  5318280 284.1  10272058   548.6    NA       10272058  548.6\n",
       "Vcells 51950130 396.4 524288982  4000.1 16384      650032873 4959.4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "in global environment after rm\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'BC_subtype'</li>\n",
       "\t<li>'brca.clin.selected'</li>\n",
       "\t<li>'brca.untran.gene.tpm'</li>\n",
       "\t<li>'date'</li>\n",
       "\t<li>'deseqdata'</li>\n",
       "\t<li>'gene1_transcripts'</li>\n",
       "\t<li>'her2.ids'</li>\n",
       "\t<li>'her2.ids.final'</li>\n",
       "\t<li>'her2.tnbc.counts.df'</li>\n",
       "\t<li>'numHER2'</li>\n",
       "\t<li>'numTNBC'</li>\n",
       "\t<li>'samplesofinterest'</li>\n",
       "\t<li>'sampleTable'</li>\n",
       "\t<li>'tnbc.ids'</li>\n",
       "\t<li>'tnbc.ids.final'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'BC\\_subtype'\n",
       "\\item 'brca.clin.selected'\n",
       "\\item 'brca.untran.gene.tpm'\n",
       "\\item 'date'\n",
       "\\item 'deseqdata'\n",
       "\\item 'gene1\\_transcripts'\n",
       "\\item 'her2.ids'\n",
       "\\item 'her2.ids.final'\n",
       "\\item 'her2.tnbc.counts.df'\n",
       "\\item 'numHER2'\n",
       "\\item 'numTNBC'\n",
       "\\item 'samplesofinterest'\n",
       "\\item 'sampleTable'\n",
       "\\item 'tnbc.ids'\n",
       "\\item 'tnbc.ids.final'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'BC_subtype'\n",
       "2. 'brca.clin.selected'\n",
       "3. 'brca.untran.gene.tpm'\n",
       "4. 'date'\n",
       "5. 'deseqdata'\n",
       "6. 'gene1_transcripts'\n",
       "7. 'her2.ids'\n",
       "8. 'her2.ids.final'\n",
       "9. 'her2.tnbc.counts.df'\n",
       "10. 'numHER2'\n",
       "11. 'numTNBC'\n",
       "12. 'samplesofinterest'\n",
       "13. 'sampleTable'\n",
       "14. 'tnbc.ids'\n",
       "15. 'tnbc.ids.final'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"BC_subtype\"           \"brca.clin.selected\"   \"brca.untran.gene.tpm\"\n",
       " [4] \"date\"                 \"deseqdata\"            \"gene1_transcripts\"   \n",
       " [7] \"her2.ids\"             \"her2.ids.final\"       \"her2.tnbc.counts.df\" \n",
       "[10] \"numHER2\"              \"numTNBC\"              \"samplesofinterest\"   \n",
       "[13] \"sampleTable\"          \"tnbc.ids\"             \"tnbc.ids.final\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>limit (Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Ncells</th><td> 5318307 </td><td>284.1    </td><td> 10272058</td><td> 548.6   </td><td>   NA    </td><td> 10272058</td><td> 548.6   </td></tr>\n",
       "\t<tr><th scope=row>Vcells</th><td>51950146 </td><td>396.4    </td><td>419431185</td><td>3200.1   </td><td>16384    </td><td>650032873</td><td>4959.4   </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllll}\n",
       "  & used & (Mb) & gc trigger & (Mb) & limit (Mb) & max used & (Mb)\\\\\n",
       "\\hline\n",
       "\tNcells &  5318307  & 284.1     &  10272058 &  548.6    &    NA     &  10272058 &  548.6   \\\\\n",
       "\tVcells & 51950146  & 396.4     & 419431185 & 3200.1    & 16384     & 650032873 & 4959.4   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | used | (Mb) | gc trigger | (Mb) | limit (Mb) | max used | (Mb) | \n",
       "|---|---|\n",
       "| Ncells |  5318307  | 284.1     |  10272058 |  548.6    |    NA     |  10272058 |  548.6    | \n",
       "| Vcells | 51950146  | 396.4     | 419431185 | 3200.1    | 16384     | 650032873 | 4959.4    | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "       used     (Mb)  gc trigger (Mb)   limit (Mb) max used  (Mb)  \n",
       "Ncells  5318307 284.1  10272058   548.6    NA       10272058  548.6\n",
       "Vcells 51950146 396.4 419431185  3200.1 16384      650032873 4959.4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# to view what is in the global environment before clearing\n",
    "message(\"in global environment before rm\")\n",
    "ls()\n",
    "\n",
    "#remove large elements that are not needed. note we need tissuesource, samplesofinterest, and brca.untran.gene.tpm.df\n",
    "rm(brca.untran.counts.df, brca.untran.gene.counts, her2.counts.df, brca.untran.tpm.df, df.txi, tx2gene81, tnbc.counts.df, txi)\n",
    "gc()\n",
    "\n",
    "#to make sure it worked\n",
    "message(\"in global environment after rm\")\n",
    "ls()\n",
    "\n",
    "#to free up ram\n",
    "gc()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now we are ready to perform the statistical analysis. For the purposes of this exercise we will remove genes that are expressed at a low level across many samples to drastically increase the speed of the analysis. However at the end of this notebook, I include the code  (all commented out) that I normally would run that allows DESeq2 to perform a statistical test across all genes instead of arbitrary filtering. \n",
    "Note that we use the SnowParam function to make the code run across multiple threads. The number of threads depends on your machine and if you want to dedicate it completely to this task. On a relatively new laptop this step should not take more than a few hours. However, you might want to run overnight.\n",
    "\n",
    "The verbose output should read something like:\n",
    "```\n",
    "using pre-existing size factors\n",
    "estimating dispersions\n",
    "gene-wise dispersion estimates: 4 workers\n",
    "mean-dispersion relationship\n",
    "final dispersion estimates, fitting model and testing: 4 workers\n",
    "-- replacing outliers and refitting for 502 genes\n",
    "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
    "-- original counts are preserved in counts(dds)\n",
    "estimating dispersions\n",
    "fitting model and testing\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "using pre-existing size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 4 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 4 workers\n",
      "-- replacing outliers and refitting for 502 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "estimating dispersions\n",
      "fitting model and testing\n"
     ]
    }
   ],
   "source": [
    "##################################################################################################\n",
    "################ code to execute DEseq2 after removing lowly expressed genes######################\n",
    "################ this is only done to decrease computational time.  ###############################\n",
    "############### normally run the code above and would skip this     ###############################\n",
    "###################################################################################################\n",
    "\n",
    "#will make DESeq2 run across 4 threads. You can change the number if running on a machine/server with more or less than 4.\n",
    "numCores <- 4\n",
    "register(SnowParam(numCores))\n",
    "\n",
    "\n",
    "#these commands remove genes with low counts\n",
    "#the first number specifies the number of counts for a given gene necessary to include in the deseq analysis\n",
    "#the second number tells how many samples need to have that mean counts\n",
    "#so in this case only genes with at least 200 counts in 100 samples will be kept\n",
    "#this is much higher than I would normally do\n",
    "#these numbers were selected only to reduce computational time for this exercise\n",
    "#Normally, filtering is not necessary before running\n",
    "dds <- estimateSizeFactors(deseqdata)\n",
    "idx <- rowSums(counts(dds, normalized=TRUE) >= 200 ) >= 100\n",
    "dds <- dds[idx,]\n",
    "\n",
    "#this is the DESeq execution command that will take ~1hr\n",
    "dds <- DESeq(dds, parallel = TRUE)\n",
    "\n",
    "#save after running to save computational time later\n",
    "save(dds, file = file.path(date, \"her2.tnbc.results.Rda\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create an output directory to save all tables and graphs and make the differential expression results readable to any user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000141736</th><td>111881.653   </td><td>-3.907757    </td><td>0.1497027    </td><td>-26.10345    </td><td>3.331450e-150</td><td>4.932878e-146</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000161395</th><td> 10474.829   </td><td>-3.289347    </td><td>0.1397486    </td><td>-23.53760    </td><td>1.681954e-122</td><td>1.245234e-118</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000171428</th><td>  4752.450   </td><td>-4.556832    </td><td>0.2036122    </td><td>-22.37996    </td><td>6.170364e-111</td><td>3.045486e-107</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000105388</th><td>  4442.615   </td><td>-5.409955    </td><td>0.2490188    </td><td>-21.72509    </td><td>1.188649e-104</td><td>4.400083e-101</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000108932</th><td>  1935.616   </td><td>-3.610246    </td><td>0.1665687    </td><td>-21.67421    </td><td>3.593392e-104</td><td>1.064147e-100</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000172057</th><td> 18939.110   </td><td>-2.936965    </td><td>0.1386755    </td><td>-21.17869    </td><td> 1.501635e-99</td><td> 3.705785e-96</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llllll}\n",
       "  & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj\\\\\n",
       "\\hline\n",
       "\tENSG00000141736 & 111881.653    & -3.907757     & 0.1497027     & -26.10345     & 3.331450e-150 & 4.932878e-146\\\\\n",
       "\tENSG00000161395 &  10474.829    & -3.289347     & 0.1397486     & -23.53760     & 1.681954e-122 & 1.245234e-118\\\\\n",
       "\tENSG00000171428 &   4752.450    & -4.556832     & 0.2036122     & -22.37996     & 6.170364e-111 & 3.045486e-107\\\\\n",
       "\tENSG00000105388 &   4442.615    & -5.409955     & 0.2490188     & -21.72509     & 1.188649e-104 & 4.400083e-101\\\\\n",
       "\tENSG00000108932 &   1935.616    & -3.610246     & 0.1665687     & -21.67421     & 3.593392e-104 & 1.064147e-100\\\\\n",
       "\tENSG00000172057 &  18939.110    & -2.936965     & 0.1386755     & -21.17869     &  1.501635e-99 &  3.705785e-96\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | baseMean | log2FoldChange | lfcSE | stat | pvalue | padj | \n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000141736 | 111881.653    | -3.907757     | 0.1497027     | -26.10345     | 3.331450e-150 | 4.932878e-146 | \n",
       "| ENSG00000161395 |  10474.829    | -3.289347     | 0.1397486     | -23.53760     | 1.681954e-122 | 1.245234e-118 | \n",
       "| ENSG00000171428 |   4752.450    | -4.556832     | 0.2036122     | -22.37996     | 6.170364e-111 | 3.045486e-107 | \n",
       "| ENSG00000105388 |   4442.615    | -5.409955     | 0.2490188     | -21.72509     | 1.188649e-104 | 4.400083e-101 | \n",
       "| ENSG00000108932 |   1935.616    | -3.610246     | 0.1665687     | -21.67421     | 3.593392e-104 | 1.064147e-100 | \n",
       "| ENSG00000172057 |  18939.110    | -2.936965     | 0.1386755     | -21.17869     |  1.501635e-99 |  3.705785e-96 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "                baseMean   log2FoldChange lfcSE     stat      pvalue       \n",
       "ENSG00000141736 111881.653 -3.907757      0.1497027 -26.10345 3.331450e-150\n",
       "ENSG00000161395  10474.829 -3.289347      0.1397486 -23.53760 1.681954e-122\n",
       "ENSG00000171428   4752.450 -4.556832      0.2036122 -22.37996 6.170364e-111\n",
       "ENSG00000105388   4442.615 -5.409955      0.2490188 -21.72509 1.188649e-104\n",
       "ENSG00000108932   1935.616 -3.610246      0.1665687 -21.67421 3.593392e-104\n",
       "ENSG00000172057  18939.110 -2.936965      0.1386755 -21.17869  1.501635e-99\n",
       "                padj         \n",
       "ENSG00000141736 4.932878e-146\n",
       "ENSG00000161395 1.245234e-118\n",
       "ENSG00000171428 3.045486e-107\n",
       "ENSG00000105388 4.400083e-101\n",
       "ENSG00000108932 1.064147e-100\n",
       "ENSG00000172057  3.705785e-96"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#make an output director\n",
    "outdir <- \"DESeq_output\"\n",
    "dir.create(file.path(date, outdir), recursive = TRUE)\n",
    "\n",
    "\n",
    "#the results function summarizes the deseq results\n",
    "#anythign with an alpha -FDR- greater than 0.001 is not considered significant\n",
    "#you can obviously change this threshold so that everything is included or to be\n",
    "#more stringent\n",
    "results.fdr.threshold <- results(dds, alpha = 0.001, parallel = TRUE)\n",
    "save(results.fdr.threshold, file = file.path(date, outdir, \"her2.tnbc.results.Rda\"))\n",
    "\n",
    "#make it a dataframe so that we can export it to a csv\n",
    "results.fdr.threshold.df <- as.data.frame(results.fdr.threshold)\n",
    "\n",
    "#reorder based on pvalue\n",
    "results.ordered <- results.fdr.threshold.df[order(results.fdr.threshold.df$pvalue, na.last = TRUE),]\n",
    "\n",
    "write.csv(results.ordered, file = file.path(date, outdir, \"her2.tnbc.results.csv\"))\n",
    "\n",
    "#check what results file looks like\n",
    "head(results.ordered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print to screen simple summary statistics\n",
    "In R, I save the summary below to file by using the sink command that is not compatible in Jupyter Notebooks\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "summary of dds\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "out of 14807 with nonzero total read count\n",
      "adjusted p-value < 0.001\n",
      "LFC > 0 (up)       : 3996, 27%\n",
      "LFC < 0 (down)     : 3375, 23%\n",
      "outliers [1]       : 0, 0%\n",
      "low counts [2]     : 0, 0%\n",
      "(mean count < 150)\n",
      "[1] see 'cooksCutoff' argument of ?results\n",
      "[2] see 'independentFiltering' argument of ?results\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "summary of data table\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "    baseMean        log2FoldChange         lfcSE              stat         \n",
       " Min.   :   149.6   Min.   :-6.80133   Min.   :0.02868   Min.   :-26.1034  \n",
       " 1st Qu.:   774.4   1st Qu.:-0.24955   1st Qu.:0.06541   1st Qu.: -3.0945  \n",
       " Median :  1870.1   Median : 0.04124   Median :0.08384   Median :  0.5140  \n",
       " Mean   :  4510.8   Mean   : 0.03333   Mean   :0.09918   Mean   :  0.2575  \n",
       " 3rd Qu.:  4034.2   3rd Qu.: 0.33513   3rd Qu.:0.11492   3rd Qu.:  3.7944  \n",
       " Max.   :789875.0   Max.   : 5.86999   Max.   :0.59752   Max.   : 17.9066  \n",
       "     pvalue               padj         \n",
       " Min.   :0.0000000   Min.   :0.000000  \n",
       " 1st Qu.:0.0000000   1st Qu.:0.000000  \n",
       " Median :0.0005185   Median :0.001037  \n",
       " Mean   :0.1226463   Mean   :0.136650  \n",
       " 3rd Qu.:0.1027516   3rd Qu.:0.136999  \n",
       " Max.   :0.9998667   Max.   :0.999867  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#print summary of deseq to screen. \n",
    "#in native R, save this output by uncommenting the sink commands\n",
    "\n",
    "\n",
    "#sink(file.path(date, \"DESeqsummary_HER2_TNBC.txt\"))\n",
    "#note this is a summary  of the results BEFORE making it a dataframe\n",
    "message(\"summary of dds\")\n",
    "summary(results.fdr.threshold)\n",
    "#sink()\n",
    "\n",
    "#this is a summary of the dataframe\n",
    "#notice the difference in the summary statistics\n",
    "message(\"summary of data table\")\n",
    "summary(results.ordered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**note** this provides  a lot of information. First, there are only 14,807 genes with non-zero counts. This means we filtered a lot out before the differential expressional analysis. This number will be different if you run the code at the end of this notebook that do not exclude genes expressed at low levels. This summary also tells us that there are 3996 genes significantly upregulated in TNBC and 3375 significantly downregulated (FDR <0.001). \n",
    "\n",
    "The summary stats of the data table also tell us a great deal. The log2FCs range from -6.8 to 5.87. The padj column tells us that at least a quarter of the genes have an FDR of <0.000001 meaning the HER2 and TNBC groups are very different."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The differential expression statistical analysis is complete.\n",
    "In the next notebook, we will annotate the results file and visually explore the data.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "___\n",
    "___\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Below is the code to execute the differential expression analysis without ignoring genes expressed at low levels. Without a time restriction, I recommend doing this because arbitrary cutoffs will lead to false negatives."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################################################\n",
    "################### execute DESeq2 #####################################################\n",
    "#########################################################################################\n",
    "\n",
    "#Normally I would run everything below, but this will take ~4 hours on 4 threads on this server\n",
    "#accordingly, run the next code set that will remove genes with low expression, which will decrease\n",
    "#the computational time to ~1hr.\n",
    "\n",
    "\n",
    "load(file = file.path(date,\"her2vstnbc.deseq.Rda\"))\n",
    "\n",
    "#will run across 4 threads\n",
    "numCores <- 4\n",
    "register(SnowParam(numCores))\n",
    "\n",
    "dds.total <- DESeq(deseqdata, parallel = TRUE)\n",
    "save(dds.total, file = \"her2.dds.total.Rda\")\n",
    "Sys.time()\n",
    "\n",
    "fdr = .001\n",
    "results.total <- results(dds.total, alpha = fdr, parallel = TRUE)\n",
    "save(results.total, file = file.path(date, \"her2.tnbc.total.results.Rda\"))\n",
    "\n",
    "#reorder based on pvalue\n",
    "results.total.df <- as.data.frame(results.total)\n",
    "results.total.ordered <- results.total.df[order(results.total.df$pvalue, na.last = TRUE),]\n",
    "\n",
    "write.csv(results.total.ordered, file = file.path(date, outdir, \"her2.tnbc.results.total.ordered.csv\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
